// @wordpress/block-library/build/search/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isPercentageUnit = isPercentageUnit;
exports.CSS_UNITS = exports.MIN_WIDTH_UNIT = exports.MIN_WIDTH = exports.PX_WIDTH_DEFAULT = exports.PC_WIDTH_DEFAULT = void 0;

var _i18n = require("@wordpress/i18n");

var _element = require("@wordpress/element");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Constants
 */
var isWeb = _element.Platform.OS === 'web';
var PC_WIDTH_DEFAULT = 50;
exports.PC_WIDTH_DEFAULT = PC_WIDTH_DEFAULT;
var PX_WIDTH_DEFAULT = 350;
exports.PX_WIDTH_DEFAULT = PX_WIDTH_DEFAULT;
var MIN_WIDTH = 220;
exports.MIN_WIDTH = MIN_WIDTH;
var MIN_WIDTH_UNIT = 'px';
exports.MIN_WIDTH_UNIT = MIN_WIDTH_UNIT;
var CSS_UNITS = [{
  value: '%',
  label: isWeb ? '%' : (0, _i18n.__)('Percentage (%)'),
  default: PC_WIDTH_DEFAULT
}, {
  value: 'px',
  label: isWeb ? 'px' : (0, _i18n.__)('Pixels (px)'),
  default: PX_WIDTH_DEFAULT
}];
/**
 * Returns a boolean whether passed unit is percentage
 *
 * @param {string} unit Block width unit.
 *
 * @return {boolean} 	Whether unit is '%'.
 */

exports.CSS_UNITS = CSS_UNITS;

function isPercentageUnit(unit) {
  return unit === '%';
}
