// @wordpress/block-library/build-module/navigation/block-navigation-list.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __experimentalBlockNavigationTree, store as blockEditorStore } from '@wordpress/block-editor';
import { useSelect, useDispatch } from '@wordpress/data';
export default function BlockNavigationList(_ref) {
  var clientId = _ref.clientId,
      __experimentalFeatures = _ref.__experimentalFeatures;

  var _useSelect = useSelect(function (select) {
    var _select = select(blockEditorStore),
        getSelectedBlockClientId = _select.getSelectedBlockClientId,
        __unstableGetClientIdsTree = _select.__unstableGetClientIdsTree;

    return {
      blocks: __unstableGetClientIdsTree(clientId),
      selectedBlockClientId: getSelectedBlockClientId()
    };
  }, [clientId]),
      blocks = _useSelect.blocks,
      selectedBlockClientId = _useSelect.selectedBlockClientId;

  var _useDispatch = useDispatch(blockEditorStore),
      selectBlock = _useDispatch.selectBlock;

  return createElement(__experimentalBlockNavigationTree, {
    blocks: blocks,
    selectedBlockClientIds: [selectedBlockClientId],
    selectBlock: selectBlock,
    __experimentalFeatures: __experimentalFeatures,
    showNestedBlocks: true,
    showAppender: true,
    showBlockMovers: true
  });
}
