// @wordpress/block-library/build-module/navigation/edit.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { useState } from '@wordpress/element';
import { InnerBlocks, __experimentalUseInnerBlocksProps as useInnerBlocksProps, InspectorControls, JustifyToolbar, BlockControls, useBlockProps, store as blockEditorStore } from '@wordpress/block-editor';
import { useDispatch, withSelect, withDispatch } from '@wordpress/data';
import { PanelBody, ToggleControl, ToolbarGroup } from '@wordpress/components';
import { compose } from '@wordpress/compose';
import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import useBlockNavigator from './use-block-navigator';
import NavigationPlaceholder from './placeholder';
import PlaceholderPreview from './placeholder-preview';
var ALLOWED_BLOCKS = ['core/navigation-link', 'core/search', 'core/social-links', 'core/page-list', 'core/spacer'];
var LAYOUT = {
  type: 'default',
  alignments: []
};

function Navigation(_ref) {
  var _classnames;

  var selectedBlockHasDescendants = _ref.selectedBlockHasDescendants,
      attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      clientId = _ref.clientId,
      hasExistingNavItems = _ref.hasExistingNavItems,
      isImmediateParentOfSelectedBlock = _ref.isImmediateParentOfSelectedBlock,
      isSelected = _ref.isSelected,
      updateInnerBlocks = _ref.updateInnerBlocks,
      className = _ref.className,
      _ref$hasSubmenuIndica = _ref.hasSubmenuIndicatorSetting,
      hasSubmenuIndicatorSetting = _ref$hasSubmenuIndica === void 0 ? true : _ref$hasSubmenuIndica,
      _ref$hasItemJustifica = _ref.hasItemJustificationControls,
      hasItemJustificationControls = _ref$hasItemJustifica === void 0 ? true : _ref$hasItemJustifica;

  var _useState = useState(!hasExistingNavItems),
      _useState2 = _slicedToArray(_useState, 2),
      isPlaceholderShown = _useState2[0],
      setIsPlaceholderShown = _useState2[1];

  var _useDispatch = useDispatch(blockEditorStore),
      selectBlock = _useDispatch.selectBlock;

  var blockProps = useBlockProps({
    className: classnames(className, (_classnames = {}, _defineProperty(_classnames, "items-justified-".concat(attributes.itemsJustification), attributes.itemsJustification), _defineProperty(_classnames, 'is-vertical', attributes.orientation === 'vertical'), _classnames))
  });

  var _useBlockNavigator = useBlockNavigator(clientId),
      navigatorToolbarButton = _useBlockNavigator.navigatorToolbarButton,
      navigatorModal = _useBlockNavigator.navigatorModal;

  var innerBlocksProps = useInnerBlocksProps({
    className: 'wp-block-navigation__container'
  }, {
    allowedBlocks: ALLOWED_BLOCKS,
    orientation: attributes.orientation || 'horizontal',
    renderAppender: isImmediateParentOfSelectedBlock && !selectedBlockHasDescendants || isSelected ? InnerBlocks.DefaultAppender : false,
    __experimentalAppenderTagName: 'li',
    __experimentalCaptureToolbars: true,
    // Template lock set to false here so that the Nav
    // Block on the experimental menus screen does not
    // inherit templateLock={ 'all' }.
    templateLock: false,
    __experimentalLayout: LAYOUT,
    placeholder: createElement(PlaceholderPreview, null)
  });

  if (isPlaceholderShown) {
    return createElement("div", blockProps, createElement(NavigationPlaceholder, {
      onCreate: function onCreate(blocks, selectNavigationBlock) {
        setIsPlaceholderShown(false);
        updateInnerBlocks(blocks);

        if (selectNavigationBlock) {
          selectBlock(clientId);
        }
      }
    }));
  }

  var justifyAllowedControls = attributes.orientation === 'vertical' ? ['left', 'center', 'right'] : ['left', 'center', 'right', 'space-between'];
  return createElement(Fragment, null, createElement(BlockControls, null, hasItemJustificationControls && createElement(JustifyToolbar, {
    value: attributes.itemsJustification,
    allowedControls: justifyAllowedControls,
    onChange: function onChange(value) {
      return setAttributes({
        itemsJustification: value
      });
    },
    popoverProps: {
      position: 'bottom right',
      isAlternate: true
    }
  }), createElement(ToolbarGroup, null, navigatorToolbarButton)), navigatorModal, createElement(InspectorControls, null, hasSubmenuIndicatorSetting && createElement(PanelBody, {
    title: __('Display settings')
  }, createElement(ToggleControl, {
    checked: attributes.showSubmenuIcon,
    onChange: function onChange(value) {
      setAttributes({
        showSubmenuIcon: value
      });
    },
    label: __('Show submenu indicator icons')
  }))), createElement("nav", blockProps, createElement("ul", innerBlocksProps)));
}

export default compose([withSelect(function (select, _ref2) {
  var _getClientIdsOfDescen;

  var clientId = _ref2.clientId;
  var innerBlocks = select(blockEditorStore).getBlocks(clientId);

  var _select = select(blockEditorStore),
      getClientIdsOfDescendants = _select.getClientIdsOfDescendants,
      hasSelectedInnerBlock = _select.hasSelectedInnerBlock,
      getSelectedBlockClientId = _select.getSelectedBlockClientId;

  var isImmediateParentOfSelectedBlock = hasSelectedInnerBlock(clientId, false);
  var selectedBlockId = getSelectedBlockClientId();
  var selectedBlockHasDescendants = !!((_getClientIdsOfDescen = getClientIdsOfDescendants([selectedBlockId])) !== null && _getClientIdsOfDescen !== void 0 && _getClientIdsOfDescen.length);
  return {
    isImmediateParentOfSelectedBlock: isImmediateParentOfSelectedBlock,
    selectedBlockHasDescendants: selectedBlockHasDescendants,
    hasExistingNavItems: !!innerBlocks.length
  };
}), withDispatch(function (dispatch, _ref3) {
  var clientId = _ref3.clientId;
  return {
    updateInnerBlocks: function updateInnerBlocks(blocks) {
      if ((blocks === null || blocks === void 0 ? void 0 : blocks.length) === 0) {
        return false;
      }

      dispatch(blockEditorStore).replaceInnerBlocks(clientId, blocks, true);
    }
  };
})])(Navigation);
