// @wordpress/block-library/build-module/navigation/use-block-navigator.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useState } from '@wordpress/element';
import { ToolbarButton, Modal } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { listView } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import BlockNavigationList from './block-navigation-list';
export default function useBlockNavigator(clientId, __experimentalFeatures) {
  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isNavigationListOpen = _useState2[0],
      setIsNavigationListOpen = _useState2[1];

  var navigatorToolbarButton = createElement(ToolbarButton, {
    className: "components-toolbar__control",
    label: __('Open block navigation'),
    onClick: function onClick() {
      return setIsNavigationListOpen(true);
    },
    icon: listView
  });
  var navigatorModal = isNavigationListOpen && createElement(Modal, {
    title: __('Navigation'),
    closeLabel: __('Close'),
    onRequestClose: function onRequestClose() {
      setIsNavigationListOpen(false);
    }
  }, createElement(BlockNavigationList, {
    clientId: clientId,
    __experimentalFeatures: __experimentalFeatures
  }));
  return {
    navigatorToolbarButton: navigatorToolbarButton,
    navigatorModal: navigatorModal
  };
}
