// @wordpress/block-library/build/navigation/block-navigation-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockNavigationList;

var _element = require("@wordpress/element");

var _blockEditor = require("@wordpress/block-editor");

var _data = require("@wordpress/data");

/**
 * WordPress dependencies
 */
function BlockNavigationList(_ref) {
  var clientId = _ref.clientId,
      __experimentalFeatures = _ref.__experimentalFeatures;

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select(_blockEditor.store),
        getSelectedBlockClientId = _select.getSelectedBlockClientId,
        __unstableGetClientIdsTree = _select.__unstableGetClientIdsTree;

    return {
      blocks: __unstableGetClientIdsTree(clientId),
      selectedBlockClientId: getSelectedBlockClientId()
    };
  }, [clientId]),
      blocks = _useSelect.blocks,
      selectedBlockClientId = _useSelect.selectedBlockClientId;

  var _useDispatch = (0, _data.useDispatch)(_blockEditor.store),
      selectBlock = _useDispatch.selectBlock;

  return (0, _element.createElement)(_blockEditor.__experimentalBlockNavigationTree, {
    blocks: blocks,
    selectedBlockClientIds: [selectedBlockClientId],
    selectBlock: selectBlock,
    __experimentalFeatures: __experimentalFeatures,
    showNestedBlocks: true,
    showAppender: true,
    showBlockMovers: true
  });
}
