// @wordpress/block-library/build/navigation/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _blockEditor = require("@wordpress/block-editor");

var _data = require("@wordpress/data");

var _components = require("@wordpress/components");

var _compose = require("@wordpress/compose");

var _i18n = require("@wordpress/i18n");

var _useBlockNavigator2 = _interopRequireDefault(require("./use-block-navigator"));

var _placeholder = _interopRequireDefault(require("./placeholder"));

var _placeholderPreview = _interopRequireDefault(require("./placeholder-preview"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var ALLOWED_BLOCKS = ['core/navigation-link', 'core/search', 'core/social-links', 'core/page-list', 'core/spacer'];
var LAYOUT = {
  type: 'default',
  alignments: []
};

function Navigation(_ref) {
  var _classnames;

  var selectedBlockHasDescendants = _ref.selectedBlockHasDescendants,
      attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      clientId = _ref.clientId,
      hasExistingNavItems = _ref.hasExistingNavItems,
      isImmediateParentOfSelectedBlock = _ref.isImmediateParentOfSelectedBlock,
      isSelected = _ref.isSelected,
      updateInnerBlocks = _ref.updateInnerBlocks,
      className = _ref.className,
      _ref$hasSubmenuIndica = _ref.hasSubmenuIndicatorSetting,
      hasSubmenuIndicatorSetting = _ref$hasSubmenuIndica === void 0 ? true : _ref$hasSubmenuIndica,
      _ref$hasItemJustifica = _ref.hasItemJustificationControls,
      hasItemJustificationControls = _ref$hasItemJustifica === void 0 ? true : _ref$hasItemJustifica;

  var _useState = (0, _element.useState)(!hasExistingNavItems),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isPlaceholderShown = _useState2[0],
      setIsPlaceholderShown = _useState2[1];

  var _useDispatch = (0, _data.useDispatch)(_blockEditor.store),
      selectBlock = _useDispatch.selectBlock;

  var blockProps = (0, _blockEditor.useBlockProps)({
    className: (0, _classnames2.default)(className, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "items-justified-".concat(attributes.itemsJustification), attributes.itemsJustification), (0, _defineProperty2.default)(_classnames, 'is-vertical', attributes.orientation === 'vertical'), _classnames))
  });

  var _useBlockNavigator = (0, _useBlockNavigator2.default)(clientId),
      navigatorToolbarButton = _useBlockNavigator.navigatorToolbarButton,
      navigatorModal = _useBlockNavigator.navigatorModal;

  var innerBlocksProps = (0, _blockEditor.__experimentalUseInnerBlocksProps)({
    className: 'wp-block-navigation__container'
  }, {
    allowedBlocks: ALLOWED_BLOCKS,
    orientation: attributes.orientation || 'horizontal',
    renderAppender: isImmediateParentOfSelectedBlock && !selectedBlockHasDescendants || isSelected ? _blockEditor.InnerBlocks.DefaultAppender : false,
    __experimentalAppenderTagName: 'li',
    __experimentalCaptureToolbars: true,
    // Template lock set to false here so that the Nav
    // Block on the experimental menus screen does not
    // inherit templateLock={ 'all' }.
    templateLock: false,
    __experimentalLayout: LAYOUT,
    placeholder: (0, _element.createElement)(_placeholderPreview.default, null)
  });

  if (isPlaceholderShown) {
    return (0, _element.createElement)("div", blockProps, (0, _element.createElement)(_placeholder.default, {
      onCreate: function onCreate(blocks, selectNavigationBlock) {
        setIsPlaceholderShown(false);
        updateInnerBlocks(blocks);

        if (selectNavigationBlock) {
          selectBlock(clientId);
        }
      }
    }));
  }

  var justifyAllowedControls = attributes.orientation === 'vertical' ? ['left', 'center', 'right'] : ['left', 'center', 'right', 'space-between'];
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.BlockControls, null, hasItemJustificationControls && (0, _element.createElement)(_blockEditor.JustifyToolbar, {
    value: attributes.itemsJustification,
    allowedControls: justifyAllowedControls,
    onChange: function onChange(value) {
      return setAttributes({
        itemsJustification: value
      });
    },
    popoverProps: {
      position: 'bottom right',
      isAlternate: true
    }
  }), (0, _element.createElement)(_components.ToolbarGroup, null, navigatorToolbarButton)), navigatorModal, (0, _element.createElement)(_blockEditor.InspectorControls, null, hasSubmenuIndicatorSetting && (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Display settings')
  }, (0, _element.createElement)(_components.ToggleControl, {
    checked: attributes.showSubmenuIcon,
    onChange: function onChange(value) {
      setAttributes({
        showSubmenuIcon: value
      });
    },
    label: (0, _i18n.__)('Show submenu indicator icons')
  }))), (0, _element.createElement)("nav", blockProps, (0, _element.createElement)("ul", innerBlocksProps)));
}

var _default = (0, _compose.compose)([(0, _data.withSelect)(function (select, _ref2) {
  var _getClientIdsOfDescen;

  var clientId = _ref2.clientId;
  var innerBlocks = select(_blockEditor.store).getBlocks(clientId);

  var _select = select(_blockEditor.store),
      getClientIdsOfDescendants = _select.getClientIdsOfDescendants,
      hasSelectedInnerBlock = _select.hasSelectedInnerBlock,
      getSelectedBlockClientId = _select.getSelectedBlockClientId;

  var isImmediateParentOfSelectedBlock = hasSelectedInnerBlock(clientId, false);
  var selectedBlockId = getSelectedBlockClientId();
  var selectedBlockHasDescendants = !!((_getClientIdsOfDescen = getClientIdsOfDescendants([selectedBlockId])) !== null && _getClientIdsOfDescen !== void 0 && _getClientIdsOfDescen.length);
  return {
    isImmediateParentOfSelectedBlock: isImmediateParentOfSelectedBlock,
    selectedBlockHasDescendants: selectedBlockHasDescendants,
    hasExistingNavItems: !!innerBlocks.length
  };
}), (0, _data.withDispatch)(function (dispatch, _ref3) {
  var clientId = _ref3.clientId;
  return {
    updateInnerBlocks: function updateInnerBlocks(blocks) {
      if ((blocks === null || blocks === void 0 ? void 0 : blocks.length) === 0) {
        return false;
      }

      dispatch(_blockEditor.store).replaceInnerBlocks(clientId, blocks, true);
    }
  };
})])(Navigation);

exports.default = _default;
