// @wordpress/block-library/build/navigation/use-block-navigator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useBlockNavigator;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _blockNavigationList = _interopRequireDefault(require("./block-navigation-list"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function useBlockNavigator(clientId, __experimentalFeatures) {
  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isNavigationListOpen = _useState2[0],
      setIsNavigationListOpen = _useState2[1];

  var navigatorToolbarButton = (0, _element.createElement)(_components.ToolbarButton, {
    className: "components-toolbar__control",
    label: (0, _i18n.__)('Open block navigation'),
    onClick: function onClick() {
      return setIsNavigationListOpen(true);
    },
    icon: _icons.listView
  });
  var navigatorModal = isNavigationListOpen && (0, _element.createElement)(_components.Modal, {
    title: (0, _i18n.__)('Navigation'),
    closeLabel: (0, _i18n.__)('Close'),
    onRequestClose: function onRequestClose() {
      setIsNavigationListOpen(false);
    }
  }, (0, _element.createElement)(_blockNavigationList.default, {
    clientId: clientId,
    __experimentalFeatures: __experimentalFeatures
  }));
  return {
    navigatorToolbarButton: navigatorToolbarButton,
    navigatorModal: navigatorModal
  };
}
