// @wordpress/block-library/build/template-part/edit/placeholder/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TemplatePartPlaceholder;

var _element = require("@wordpress/element");

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

var _blocks = require("@wordpress/blocks");

var _coreData = require("@wordpress/core-data");

var _selection = _interopRequireDefault(require("../selection"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function TemplatePartPlaceholder(_ref) {
  var area = _ref.area,
      setAttributes = _ref.setAttributes,
      innerBlocks = _ref.innerBlocks;

  var _useDispatch = (0, _data.useDispatch)(_coreData.store),
      saveEntityRecord = _useDispatch.saveEntityRecord;

  var onCreate = (0, _element.useCallback)( /*#__PURE__*/(0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
    var title, record, templatePart;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            title = (0, _i18n.__)('Untitled Template Part'); // If we have `area` set from block attributes, means an exposed
            // block variation was inserted. So add this prop to the template
            // part entity on creation. Afterwards remove `area` value from
            // block attributes.

            record = {
              title: title,
              slug: 'template-part',
              content: (0, _blocks.serialize)(innerBlocks),
              // `area` is filterable on the server and defaults to `UNCATEGORIZED`
              // if provided value is not allowed.
              area: area
            };
            _context.next = 4;
            return saveEntityRecord('postType', 'wp_template_part', record);

          case 4:
            templatePart = _context.sent;
            setAttributes({
              slug: templatePart.slug,
              theme: templatePart.theme,
              area: undefined
            });

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  })), [setAttributes, area]);
  return (0, _element.createElement)(_components.Placeholder, {
    icon: _icons.blockDefault,
    label: (0, _i18n.__)('Template Part'),
    instructions: (0, _i18n.__)('Create a new template part or pick an existing one from the list.')
  }, (0, _element.createElement)(_components.Dropdown, {
    contentClassName: "wp-block-template-part__placeholder-preview-dropdown-content",
    position: "bottom right left",
    renderToggle: function renderToggle(_ref3) {
      var isOpen = _ref3.isOpen,
          onToggle = _ref3.onToggle;
      return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.Button, {
        isPrimary: true,
        onClick: onToggle,
        "aria-expanded": isOpen
      }, (0, _i18n.__)('Choose existing')), (0, _element.createElement)(_components.Button, {
        isTertiary: true,
        onClick: onCreate
      }, (0, _i18n.__)('New template part')));
    },
    renderContent: function renderContent(_ref4) {
      var onClose = _ref4.onClose;
      return (0, _element.createElement)(_selection.default, {
        setAttributes: setAttributes,
        onClose: onClose
      });
    }
  }));
}
