// @wordpress/block-library/build-module/site-title/edit/level-toolbar.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { DropdownMenu } from '@wordpress/components';
import { __, sprintf } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import LevelIcon from './level-icon';
export default function LevelControl(_ref) {
  var level = _ref.level,
      onChange = _ref.onChange;
  var allControls = [1, 2, 3, 4, 5, 6, 0].map(function (currentLevel) {
    var isActive = currentLevel === level;
    return {
      icon: createElement(LevelIcon, {
        level: currentLevel,
        isPressed: isActive
      }),
      title: currentLevel === 0 ? __('Paragraph') : // translators: %s: heading level e.g: "1", "2", "3"
      sprintf(__('Heading %d'), currentLevel),
      isActive: isActive,
      onClick: function onClick() {
        return onChange(currentLevel);
      }
    };
  });
  return createElement(DropdownMenu, {
    label: __('Change heading level'),
    icon: createElement(LevelIcon, {
      level: level
    }),
    controls: allControls,
    isToolbarButton: true
  });
}
