// @wordpress/block-library/build-module/query/edit/block-setup/layout-step.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useSelect } from '@wordpress/data';
import { useState } from '@wordpress/element';
import { store as blocksStore } from '@wordpress/blocks';
import { useInstanceId } from '@wordpress/compose';
import { BlockPreview, store as blockEditorStore } from '@wordpress/block-editor';
import { __, isRTL } from '@wordpress/i18n';
import { Button, Icon, VisuallyHidden, __unstableComposite as Composite, __unstableUseCompositeState as useCompositeState, __unstableCompositeItem as CompositeItem } from '@wordpress/components';
import { chevronRight, chevronLeft } from '@wordpress/icons';

var LayoutSetupStep = function LayoutSetupStep(_ref) {
  var blockType = _ref.blockType,
      onVariationSelect = _ref.onVariationSelect,
      onBlockPatternSelect = _ref.onBlockPatternSelect;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      showBack = _useState2[0],
      setShowBack = _useState2[1];

  var _useSelect = useSelect(function (select) {
    var _select = select(blocksStore),
        getBlockVariations = _select.getBlockVariations,
        getDefaultBlockVariation = _select.getDefaultBlockVariation;

    var _select2 = select(blockEditorStore),
        __experimentalGetPatternsByBlockTypes = _select2.__experimentalGetPatternsByBlockTypes;

    var name = blockType.name;

    var _patterns = __experimentalGetPatternsByBlockTypes(name);

    var _blockVariations = getBlockVariations(name, 'block');

    return {
      defaultVariation: getDefaultBlockVariation(name, 'block'),
      blockVariations: _blockVariations,
      patterns: _patterns,
      hasBlockVariations: !!(_blockVariations !== null && _blockVariations !== void 0 && _blockVariations.length)
    };
  }, [blockType]),
      defaultVariation = _useSelect.defaultVariation,
      blockVariations = _useSelect.blockVariations,
      patterns = _useSelect.patterns,
      hasBlockVariations = _useSelect.hasBlockVariations;

  var _useState3 = useState(!(patterns !== null && patterns !== void 0 && patterns.length) && hasBlockVariations),
      _useState4 = _slicedToArray(_useState3, 2),
      showBlockVariations = _useState4[0],
      setShowBlockVariations = _useState4[1];

  var composite = useCompositeState(); // Show nothing if block variations and block pattterns do not exist.

  if (!hasBlockVariations && !(patterns !== null && patterns !== void 0 && patterns.length)) return null;
  var showPatternsList = !showBlockVariations && !!patterns.length;
  return createElement(Fragment, null, showBack && createElement(Button, {
    className: "block-setup-block-layout-back-button",
    icon: isRTL() ? chevronRight : chevronLeft,
    isTertiary: true,
    onClick: function onClick() {
      setShowBack(false);
      setShowBlockVariations(false);
    }
  }, __('Back')), createElement(Composite, _extends({}, composite, {
    role: "listbox",
    className: "block-setup-block-layout-list__container",
    "aria-label": __('Layout list')
  }), showBlockVariations && createElement(Fragment, null, blockVariations.map(function (variation) {
    return createElement(BlockVariation, {
      key: variation.name,
      variation: variation,
      onSelect: function onSelect() {
        var nextVariation = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : defaultVariation;
        return onVariationSelect(nextVariation);
      },
      composite: composite
    });
  })), showPatternsList && createElement(Fragment, null, patterns.map(function (pattern) {
    return createElement(BlockPattern, {
      key: pattern.name,
      pattern: pattern,
      onSelect: onBlockPatternSelect,
      composite: composite
    });
  })), !showBlockVariations && hasBlockVariations && createElement(BlockVariation, {
    key: defaultVariation.name,
    variation: {
      title: __('Start empty')
    },
    onSelect: function onSelect() {
      setShowBack(true);
      setShowBlockVariations(true);
    },
    composite: composite
  })));
};

function BlockPattern(_ref2) {
  var pattern = _ref2.pattern,
      onSelect = _ref2.onSelect,
      composite = _ref2.composite;
  var viewportWidth = pattern.viewportWidth,
      blocks = pattern.blocks;
  var descriptionId = useInstanceId(BlockPattern, 'block-setup-block-layout-list__item-description');
  return createElement("div", {
    className: "block-setup-block-layout-list__list-item",
    "aria-label": pattern.title,
    "aria-describedby": pattern.description ? descriptionId : undefined
  }, createElement(CompositeItem, _extends({
    role: "option",
    as: "div"
  }, composite, {
    className: "block-setup-block-layout-list__item",
    onClick: function onClick() {
      return onSelect(blocks);
    }
  }), createElement(BlockPreview, {
    blocks: blocks,
    viewportWidth: viewportWidth
  })), createElement("div", {
    className: "block-setup-block-layout-list__item-title"
  }, pattern.title), !!pattern.description && createElement(VisuallyHidden, {
    id: descriptionId
  }, pattern.description));
}

function BlockVariation(_ref3) {
  var variation = _ref3.variation,
      onSelect = _ref3.onSelect,
      composite = _ref3.composite;
  var descriptionId = useInstanceId(BlockVariation, 'block-setup-block-layout-list__item-description');
  return createElement("div", {
    className: "block-setup-block-layout-list__list-item is-block-variation",
    "aria-label": variation.title,
    "aria-describedby": variation.description ? descriptionId : undefined
  }, createElement(CompositeItem, _extends({
    role: "option",
    as: "div"
  }, composite, {
    className: "block-setup-block-layout-list__item",
    onClick: function onClick() {
      return onSelect(variation);
    },
    label: variation.description || variation.title
  }), variation.icon && createElement("div", {
    className: "block-setup-block-layout-list__item-variation-icon"
  }, createElement(Icon, {
    icon: variation.icon,
    size: 48
  }))), createElement("div", {
    className: "block-setup-block-layout-list__item-title"
  }, variation.title), !!variation.description && createElement(VisuallyHidden, {
    id: descriptionId
  }, variation.description));
}

export default LayoutSetupStep;
