// @wordpress/block-library/build/page-list/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _icons = require("@wordpress/icons");

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit.js"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/page-list",
  category: "widgets",
  usesContext: ["textColor", "customTextColor", "backgroundColor", "customBackgroundColor", "fontSize", "customFontSize", "showSubmenuIcon", "style"],
  supports: {
    reusable: false,
    html: false
  },
  editorStyle: "wp-block-page-list-editor",
  style: "wp-block-page-list"
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Page List', 'block title'),
  description: (0, _i18n.__)('Display a list of all pages.'),
  keywords: [(0, _i18n.__)('menu'), (0, _i18n.__)('navigation')],
  icon: _icons.pages,
  example: {},
  edit: _edit.default
};
exports.settings = settings;
