// @wordpress/block-library/build-module/group/edit.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { useSelect } from '@wordpress/data';
import { InnerBlocks, useBlockProps, InspectorAdvancedControls, __experimentalUseInnerBlocksProps as useInnerBlocksProps, __experimentalUseEditorFeature as useEditorFeature, store as blockEditorStore } from '@wordpress/block-editor';
import { SelectControl } from '@wordpress/components';
import { __ } from '@wordpress/i18n';

function GroupEdit(_ref) {
  var attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      clientId = _ref.clientId;

  var _useSelect = useSelect(function (select) {
    var _getSettings;

    var _select = select(blockEditorStore),
        getBlock = _select.getBlock,
        getSettings = _select.getSettings;

    var block = getBlock(clientId);
    return {
      hasInnerBlocks: !!(block && block.innerBlocks.length),
      themeSupportsLayout: (_getSettings = getSettings()) === null || _getSettings === void 0 ? void 0 : _getSettings.supportsLayout
    };
  }, [clientId]),
      hasInnerBlocks = _useSelect.hasInnerBlocks,
      themeSupportsLayout = _useSelect.themeSupportsLayout;

  var defaultLayout = useEditorFeature('layout') || {};
  var _attributes$tagName = attributes.tagName,
      TagName = _attributes$tagName === void 0 ? 'div' : _attributes$tagName,
      templateLock = attributes.templateLock,
      _attributes$layout = attributes.layout,
      layout = _attributes$layout === void 0 ? {} : _attributes$layout;
  var usedLayout = !!layout && layout.inherit ? defaultLayout : layout;
  var contentSize = usedLayout.contentSize,
      wideSize = usedLayout.wideSize;
  var alignments = contentSize || wideSize ? ['wide', 'full'] : ['left', 'center', 'right'];
  var blockProps = useBlockProps();
  var innerBlocksProps = useInnerBlocksProps(themeSupportsLayout ? blockProps : {
    className: 'wp-block-group__inner-container'
  }, {
    templateLock: templateLock,
    renderAppender: hasInnerBlocks ? undefined : InnerBlocks.ButtonBlockAppender,
    __experimentalLayout: {
      type: 'default',
      // Find a way to inject this in the support flag code (hooks).
      alignments: themeSupportsLayout ? alignments : undefined
    }
  });
  return createElement(Fragment, null, createElement(InspectorAdvancedControls, null, createElement(SelectControl, {
    label: __('HTML element'),
    options: [{
      label: __('Default (<div>)'),
      value: 'div'
    }, {
      label: '<header>',
      value: 'header'
    }, {
      label: '<main>',
      value: 'main'
    }, {
      label: '<section>',
      value: 'section'
    }, {
      label: '<article>',
      value: 'article'
    }, {
      label: '<aside>',
      value: 'aside'
    }, {
      label: '<footer>',
      value: 'footer'
    }],
    value: TagName,
    onChange: function onChange(value) {
      return setAttributes({
        tagName: value
      });
    }
  })), themeSupportsLayout && createElement(TagName, innerBlocksProps), !themeSupportsLayout && createElement(TagName, blockProps, createElement("div", innerBlocksProps)));
}

export default GroupEdit;
