// @wordpress/block-library/build/legacy-widget/edit/preview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PreviewIframe;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _url = require("@wordpress/url");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
var DEFAULT_HEIGHT = 300;

function PreviewIframe(_ref) {
  var idBase = _ref.idBase,
      instance = _ref.instance,
      isVisible = _ref.isVisible;
  var ref = (0, _element.useRef)();

  var _useState = (0, _element.useState)(DEFAULT_HEIGHT),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      height = _useState2[0],
      setHeight = _useState2[1];

  var adjustHeight = (0, _element.useCallback)(function () {
    setHeight(ref.current.contentDocument.body.scrollHeight);
  }, []);
  (0, _element.useEffect)(function () {
    if (isVisible) {
      adjustHeight();
    }
  }, [isVisible, adjustHeight]);
  return (0, _element.createElement)(_components.Disabled, {
    hidden: !isVisible
  }, (0, _element.createElement)("iframe", {
    ref: ref,
    className: "wp-block-legacy-widget__edit-preview",
    src: (0, _url.addQueryArgs)('themes.php', {
      page: 'gutenberg-widgets',
      'legacy-widget-preview': {
        idBase: idBase,
        instance: instance
      }
    }),
    title: (0, _i18n.__)('Legacy Widget Preview'),
    height: height,
    onLoad: adjustHeight
  }));
}
