// @wordpress/block-library/build/legacy-widget/edit/widget-type-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = WidgetTypeSelector;

var _element = require("@wordpress/element");

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _coreData = require("@wordpress/core-data");

var _blockEditor = require("@wordpress/block-editor");

/**
 * WordPress dependencies
 */
function WidgetTypeSelector(_ref) {
  var selectedId = _ref.selectedId,
      onSelect = _ref.onSelect;
  var widgetTypes = (0, _data.useSelect)(function (select) {
    var _select$getSettings$w, _select$getSettings, _select$getWidgetType;

    var hiddenIds = (_select$getSettings$w = (_select$getSettings = select(_blockEditor.store).getSettings()) === null || _select$getSettings === void 0 ? void 0 : _select$getSettings.widgetTypesToHideFromLegacyWidgetBlock) !== null && _select$getSettings$w !== void 0 ? _select$getSettings$w : [];
    return (_select$getWidgetType = select(_coreData.store).getWidgetTypes({
      per_page: -1
    })) === null || _select$getWidgetType === void 0 ? void 0 : _select$getWidgetType.filter(function (widgetType) {
      return !hiddenIds.includes(widgetType.id);
    });
  }, []);

  if (!widgetTypes) {
    return (0, _element.createElement)(_components.Spinner, null);
  }

  if (widgetTypes.length === 0) {
    return (0, _i18n.__)('There are no widgets available.');
  }

  return (0, _element.createElement)(_components.SelectControl, {
    label: (0, _i18n.__)('Select a legacy widget to display:'),
    value: selectedId !== null && selectedId !== void 0 ? selectedId : '',
    options: [{
      value: '',
      label: (0, _i18n.__)('Select widget')
    }].concat((0, _toConsumableArray2.default)(widgetTypes.map(function (widgetType) {
      return {
        value: widgetType.id,
        label: widgetType.name
      };
    }))),
    onChange: function onChange(value) {
      if (value) {
        var selected = widgetTypes.find(function (widgetType) {
          return widgetType.id === value;
        });
        onSelect({
          selectedId: selected.id,
          isMulti: selected.is_multi
        });
      } else {
        onSelect({
          selectedId: null
        });
      }
    }
  });
}
