// @wordpress/block-library/build/block/edit.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ReusableBlockEdit;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _reactNative = require("react-native");

var _coreData = require("@wordpress/core-data");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

var _compose = require("@wordpress/compose");

var _icons = require("@wordpress/icons");

var _reusableBlocks = require("@wordpress/reusable-blocks");

var _editor = _interopRequireDefault(require("./editor.scss"));

var _editTitle = _interopRequireDefault(require("./edit-title"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ReusableBlockEdit(_ref) {
  var ref = _ref.attributes.ref,
      clientId = _ref.clientId,
      isSelected = _ref.isSelected;
  var recordArgs = ['postType', 'wp_block', ref];

  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      showHelp = _useState2[0],
      setShowHelp = _useState2[1];

  var infoTextStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.infoText, _editor.default.infoTextDark);
  var infoTitleStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.infoTitle, _editor.default.infoTitleDark);
  var infoSheetIconStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.infoSheetIcon, _editor.default.infoSheetIconDark);
  var spinnerStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.spinner, _editor.default.spinnerDark);

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select2, _select3;

    var _select = select(_blockEditor.store),
        getSettings = _select.getSettings;

    return {
      reusableBlock: (_select2 = select(_coreData.store)).getEditedEntityRecord.apply(_select2, recordArgs),
      hasResolved: select(_coreData.store).hasFinishedResolution('getEditedEntityRecord', recordArgs),
      isSaving: (_select3 = select(_coreData.store)).isSavingEntityRecord.apply(_select3, recordArgs),
      canUserUpdate: select(_coreData.store).canUser('update', 'blocks', ref),
      isEditing: select(_reusableBlocks.store).__experimentalIsEditingReusableBlock(clientId),
      settings: getSettings()
    };
  }, [ref, clientId]),
      reusableBlock = _useSelect.reusableBlock,
      hasResolved = _useSelect.hasResolved,
      isEditing = _useSelect.isEditing,
      settings = _useSelect.settings;

  var _useEntityBlockEditor = (0, _coreData.useEntityBlockEditor)('postType', 'wp_block', {
    id: ref
  }),
      _useEntityBlockEditor2 = (0, _slicedToArray2.default)(_useEntityBlockEditor, 3),
      blocks = _useEntityBlockEditor2[0],
      onInput = _useEntityBlockEditor2[1],
      onChange = _useEntityBlockEditor2[2];

  function openSheet() {
    setShowHelp(true);
  }

  function closeSheet() {
    setShowHelp(false);
  }

  function renderSheet() {
    var infoTitle = _reactNative.Platform.OS === 'android' ? (0, _i18n.__)("Reusable blocks aren't editable on WordPress for Android") : (0, _i18n.__)("Reusable blocks aren't editable on WordPress for iOS");
    return (0, _element.createElement)(_components.BottomSheet, {
      isVisible: showHelp,
      hideHeader: true,
      onClose: closeSheet
    }, (0, _element.createElement)(_reactNative.View, {
      style: _editor.default.infoContainer
    }, (0, _element.createElement)(_components.Icon, {
      icon: _icons.help,
      color: infoSheetIconStyle.color,
      size: _editor.default.infoSheetIcon.size
    }), (0, _element.createElement)(_reactNative.Text, {
      style: [infoTextStyle, infoTitleStyle]
    }, infoTitle)));
  }

  if (!hasResolved) {
    return (0, _element.createElement)(_reactNative.View, {
      style: spinnerStyle
    }, (0, _element.createElement)(_reactNative.ActivityIndicator, {
      animating: true
    }));
  }

  if (!reusableBlock) {
    return (0, _element.createElement)(_reactNative.Text, null, (0, _i18n.__)('Block has been deleted or is unavailable.'));
  }

  var title = reusableBlock.title;
  var element = (0, _element.createElement)(_blockEditor.BlockEditorProvider, {
    settings: settings,
    value: blocks,
    onChange: onChange,
    onInput: onInput
  }, (0, _element.createElement)(_blockEditor.BlockList, {
    withFooter: false,
    marginHorizontal: 0
  }));

  if (!isEditing) {
    element = (0, _element.createElement)(_components.Disabled, null, element);
  }

  return (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
    disabled: !isSelected,
    accessibilityLabel: (0, _i18n.__)('Help button'),
    accessibilityRole: 'button',
    accessibilityHint: (0, _i18n.__)('Tap here to show help'),
    onPress: openSheet
  }, (0, _element.createElement)(_reactNative.View, null, isSelected && (0, _element.createElement)(_editTitle.default, {
    title: title
  }), element, renderSheet()));
}
