// @wordpress/block-library/build/button/save.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = save;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _blockEditor = require("@wordpress/block-editor");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function save(_ref) {
  var _style$border, _style$typography, _classnames;

  var attributes = _ref.attributes,
      className = _ref.className;
  var fontSize = attributes.fontSize,
      linkTarget = attributes.linkTarget,
      rel = attributes.rel,
      style = attributes.style,
      text = attributes.text,
      title = attributes.title,
      url = attributes.url,
      width = attributes.width;

  if (!text) {
    return null;
  }

  var borderRadius = style === null || style === void 0 ? void 0 : (_style$border = style.border) === null || _style$border === void 0 ? void 0 : _style$border.radius;
  var colorProps = (0, _blockEditor.__experimentalGetColorClassesAndStyles)(attributes);
  var buttonClasses = (0, _classnames2.default)('wp-block-button__link', colorProps.className, {
    'no-border-radius': borderRadius === 0
  });

  var buttonStyle = _objectSpread({
    borderRadius: borderRadius ? borderRadius + 'px' : undefined
  }, colorProps.style); // The use of a `title` attribute here is soft-deprecated, but still applied
  // if it had already been assigned, for the sake of backward-compatibility.
  // A title will no longer be assigned for new or updated button block links.


  var wrapperClasses = (0, _classnames2.default)(className, (_classnames = {}, (0, _defineProperty2.default)(_classnames, "has-custom-width wp-block-button__width-".concat(width), width), (0, _defineProperty2.default)(_classnames, "has-custom-font-size", fontSize || (style === null || style === void 0 ? void 0 : (_style$typography = style.typography) === null || _style$typography === void 0 ? void 0 : _style$typography.fontSize)), _classnames));
  return (0, _element.createElement)("div", _blockEditor.useBlockProps.save({
    className: wrapperClasses
  }), (0, _element.createElement)(_blockEditor.RichText.Content, {
    tagName: "a",
    className: buttonClasses,
    href: url,
    title: title,
    style: buttonStyle,
    value: text,
    target: linkTarget,
    rel: rel
  }));
}
