// @wordpress/block-library/build/loginout/edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LoginOutEdit;

var _element = require("@wordpress/element");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

/**
 * WordPress dependencies
 */
function LoginOutEdit(_ref) {
  var attributes = _ref.attributes,
      setAttributes = _ref.setAttributes;
  var displayLoginAsForm = attributes.displayLoginAsForm,
      redirectToCurrent = attributes.redirectToCurrent;
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.InspectorControls, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Login/out settings')
  }, (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Display login as form'),
    checked: displayLoginAsForm,
    onChange: function onChange() {
      return setAttributes({
        displayLoginAsForm: !displayLoginAsForm
      });
    }
  }), (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Redirect to current URL'),
    checked: redirectToCurrent,
    onChange: function onChange() {
      return setAttributes({
        redirectToCurrent: !redirectToCurrent
      });
    }
  }))), (0, _element.createElement)("div", (0, _blockEditor.useBlockProps)({
    className: 'logged-in'
  }), (0, _element.createElement)("a", {
    href: "#login-pseudo-link"
  }, (0, _i18n.__)('Log out'))));
}
