// @wordpress/block-library/build-module/loginout/edit.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { PanelBody, ToggleControl } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { InspectorControls, useBlockProps } from '@wordpress/block-editor';
export default function LoginOutEdit(_ref) {
  var attributes = _ref.attributes,
      setAttributes = _ref.setAttributes;
  var displayLoginAsForm = attributes.displayLoginAsForm,
      redirectToCurrent = attributes.redirectToCurrent;
  return createElement(Fragment, null, createElement(InspectorControls, null, createElement(PanelBody, {
    title: __('Login/out settings')
  }, createElement(ToggleControl, {
    label: __('Display login as form'),
    checked: displayLoginAsForm,
    onChange: function onChange() {
      return setAttributes({
        displayLoginAsForm: !displayLoginAsForm
      });
    }
  }), createElement(ToggleControl, {
    label: __('Redirect to current URL'),
    checked: redirectToCurrent,
    onChange: function onChange() {
      return setAttributes({
        redirectToCurrent: !redirectToCurrent
      });
    }
  }))), createElement("div", useBlockProps({
    className: 'logged-in'
  }), createElement("a", {
    href: "#login-pseudo-link"
  }, __('Log out'))));
}
