// @wordpress/block-editor/build/components/block-switcher/pattern-transformations-menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _icons = require("@wordpress/icons");

var _components = require("@wordpress/components");

var _blockPreview = _interopRequireDefault(require("../block-preview"));

var _useTransformedPatterns = _interopRequireDefault(require("./use-transformed-patterns"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function PatternTransformationsMenu(_ref) {
  var blocks = _ref.blocks,
      statePatterns = _ref.patterns,
      onSelect = _ref.onSelect;

  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      showTransforms = _useState2[0],
      setShowTransforms = _useState2[1];

  var patterns = (0, _useTransformedPatterns.default)(statePatterns, blocks);
  if (!patterns.length) return null;
  return (0, _element.createElement)(_components.MenuGroup, {
    className: "block-editor-block-switcher__pattern__transforms__menugroup"
  }, showTransforms && (0, _element.createElement)(PreviewPatternsPopover, {
    patterns: patterns,
    onSelect: onSelect
  }), (0, _element.createElement)(_components.MenuItem, {
    onClick: function onClick(event) {
      event.preventDefault();
      setShowTransforms(!showTransforms);
    },
    icon: _icons.chevronRight
  }, (0, _i18n.__)('Patterns')));
}

function PreviewPatternsPopover(_ref2) {
  var patterns = _ref2.patterns,
      onSelect = _ref2.onSelect;
  return (0, _element.createElement)("div", {
    className: "block-editor-block-switcher__popover__preview__parent"
  }, (0, _element.createElement)("div", {
    className: "block-editor-block-switcher__popover__preview__container"
  }, (0, _element.createElement)(_components.Popover, {
    className: "block-editor-block-switcher__preview__popover",
    position: "bottom right"
  }, (0, _element.createElement)("div", {
    className: "block-editor-block-switcher__preview"
  }, (0, _element.createElement)("div", {
    className: "block-editor-block-switcher__preview-title"
  }, (0, _i18n.__)('Preview')), (0, _element.createElement)(BlockPatternsList, {
    patterns: patterns,
    onSelect: onSelect
  })))));
}

function BlockPatternsList(_ref3) {
  var patterns = _ref3.patterns,
      onSelect = _ref3.onSelect;
  var composite = (0, _components.__unstableUseCompositeState)();
  return (0, _element.createElement)(_components.__unstableComposite, (0, _extends2.default)({}, composite, {
    role: "listbox",
    className: "block-editor-block-switcher__preview-patterns-container",
    "aria-label": (0, _i18n.__)('Patterns list')
  }), patterns.map(function (pattern) {
    return (0, _element.createElement)(BlockPattern, {
      key: pattern.name,
      pattern: pattern,
      onSelect: onSelect,
      composite: composite
    });
  }));
}

function BlockPattern(_ref4) {
  var pattern = _ref4.pattern,
      onSelect = _ref4.onSelect,
      composite = _ref4.composite;
  // TODO check pattern/preview width...
  var baseClassName = 'block-editor-block-switcher__preview-patterns-container';
  var descriptionId = (0, _compose.useInstanceId)(BlockPattern, "".concat(baseClassName, "-list__item-description"));
  return (0, _element.createElement)("div", {
    className: "".concat(baseClassName, "-list__list-item"),
    "aria-label": pattern.title,
    "aria-describedby": pattern.description ? descriptionId : undefined
  }, (0, _element.createElement)(_components.__unstableCompositeItem, (0, _extends2.default)({
    role: "option",
    as: "div"
  }, composite, {
    className: "".concat(baseClassName, "-list__item"),
    onClick: function onClick() {
      return onSelect(pattern.transformedBlocks);
    }
  }), (0, _element.createElement)(_blockPreview.default, {
    blocks: pattern.transformedBlocks,
    viewportWidth: pattern.viewportWidth || 500
  }), (0, _element.createElement)("div", {
    className: "".concat(baseClassName, "-list__item-title")
  }, pattern.title)), !!pattern.description && (0, _element.createElement)(_components.VisuallyHidden, {
    id: descriptionId
  }, pattern.description));
}

var _default = PatternTransformationsMenu;
exports.default = _default;
