// @wordpress/block-editor/build/components/block-mobile-toolbar/block-actions-menu.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _reactNative = require("react-native");

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _blocks = require("@wordpress/blocks");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

var _icons = require("@wordpress/icons");

var _notices = require("@wordpress/notices");

var _moverDescription = require("../block-mover/mover-description");

var _store = require("../../store");

var _blockTransformationsMenu = _interopRequireDefault(require("../block-switcher/block-transformations-menu"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var BlockActionsMenu = function BlockActionsMenu(_ref) {
  var blockTitle = _ref.blockTitle,
      canInsertBlockType = _ref.canInsertBlockType,
      getBlocksByClientId = _ref.getBlocksByClientId,
      isEmptyDefaultBlock = _ref.isEmptyDefaultBlock,
      isFirst = _ref.isFirst,
      isLast = _ref.isLast,
      rootClientId = _ref.rootClientId,
      selectedBlockClientId = _ref.selectedBlockClientId,
      selectedBlockPossibleTransformations = _ref.selectedBlockPossibleTransformations,
      createSuccessNotice = _ref.createSuccessNotice,
      duplicateBlock = _ref.duplicateBlock,
      onMoveDown = _ref.onMoveDown,
      onMoveUp = _ref.onMoveUp,
      openGeneralSidebar = _ref.openGeneralSidebar,
      pasteBlock = _ref.pasteBlock,
      removeBlocks = _ref.removeBlocks,
      anchorNodeRef = _ref.anchorNodeRef,
      isStackedHorizontally = _ref.isStackedHorizontally,
      onDelete = _ref.onDelete,
      wrapBlockMover = _ref.wrapBlockMover,
      wrapBlockSettings = _ref.wrapBlockSettings;

  var _useState = (0, _element.useState)((0, _components.getClipboard)()),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      clipboard = _useState2[0],
      setCurrentClipboard = _useState2[1];

  var blockActionsMenuPickerRef = (0, _element.useRef)();
  var blockTransformationMenuPickerRef = (0, _element.useRef)();
  var moversOptions = {
    keys: ['icon', 'actionTitle']
  };
  var clipboardBlock = clipboard && (0, _blocks.rawHandler)({
    HTML: clipboard
  })[0];
  var isPasteEnabled = clipboardBlock && canInsertBlockType(clipboardBlock.name, rootClientId);

  var _getMoversSetup = (0, _moverDescription.getMoversSetup)(isStackedHorizontally, moversOptions),
      _getMoversSetup$actio = _getMoversSetup.actionTitle,
      backwardButtonTitle = _getMoversSetup$actio.backward,
      forwardButtonTitle = _getMoversSetup$actio.forward;

  var allOptions = {
    settings: {
      id: 'settingsOption',
      label: (0, _i18n.__)('Block settings'),
      value: 'settingsOption',
      onSelect: openGeneralSidebar
    },
    backwardButton: {
      id: 'backwardButtonOption',
      label: backwardButtonTitle,
      value: 'backwardButtonOption',
      disabled: isFirst,
      onSelect: onMoveUp
    },
    forwardButton: {
      id: 'forwardButtonOption',
      label: forwardButtonTitle,
      value: 'forwardButtonOption',
      disabled: isLast,
      onSelect: onMoveDown
    },
    delete: {
      id: 'deleteOption',
      label: (0, _i18n.__)('Remove block'),
      value: 'deleteOption',
      separated: true,
      disabled: isEmptyDefaultBlock,
      onSelect: function onSelect() {
        onDelete();
        createSuccessNotice( // translators: displayed right after the block is removed.
        (0, _i18n.__)('Block removed'));
      }
    },
    transformButton: {
      id: 'transformButtonOption',
      label: (0, _i18n.__)('Transform block…'),
      value: 'transformButtonOption',
      onSelect: function onSelect() {
        if (blockTransformationMenuPickerRef.current) {
          blockTransformationMenuPickerRef.current.presentPicker();
        }
      }
    },
    copyButton: {
      id: 'copyButtonOption',
      label: (0, _i18n.__)('Copy block'),
      value: 'copyButtonOption',
      onSelect: function onSelect() {
        var serializedBlock = (0, _blocks.serialize)(getBlocksByClientId(selectedBlockClientId));
        setCurrentClipboard(serializedBlock);
        (0, _components.setClipboard)(serializedBlock);
        createSuccessNotice( // translators: displayed right after the block is copied.
        (0, _i18n.__)('Block copied'));
      }
    },
    cutButton: {
      id: 'cutButtonOption',
      label: (0, _i18n.__)('Cut block'),
      value: 'cutButtonOption',
      onSelect: function onSelect() {
        (0, _components.setClipboard)((0, _blocks.serialize)(getBlocksByClientId(selectedBlockClientId)));
        removeBlocks(selectedBlockClientId);
        createSuccessNotice( // translators: displayed right after the block is cut.
        (0, _i18n.__)('Block cut'));
      }
    },
    pasteButton: {
      id: 'pasteButtonOption',
      label: (0, _i18n.__)('Paste block after'),
      value: 'pasteButtonOption',
      onSelect: function onSelect() {
        onPasteBlock();
        createSuccessNotice( // translators: displayed right after the block is pasted.
        (0, _i18n.__)('Block pasted'));
      }
    },
    duplicateButton: {
      id: 'duplicateButtonOption',
      label: (0, _i18n.__)('Duplicate block'),
      value: 'duplicateButtonOption',
      onSelect: function onSelect() {
        duplicateBlock();
        createSuccessNotice( // translators: displayed right after the block is duplicated.
        (0, _i18n.__)('Block duplicated'));
      }
    }
  };
  var options = (0, _lodash.compact)([wrapBlockMover && allOptions.backwardButton, wrapBlockMover && allOptions.forwardButton, wrapBlockSettings && allOptions.settings, selectedBlockPossibleTransformations.length && allOptions.transformButton, allOptions.copyButton, allOptions.cutButton, isPasteEnabled && allOptions.pasteButton, allOptions.duplicateButton, allOptions.delete]);

  function onPasteBlock() {
    if (!clipboard) {
      return;
    }

    pasteBlock((0, _blocks.rawHandler)({
      HTML: clipboard
    })[0]);
  }

  function onPickerSelect(value) {
    var selectedItem = options.find(function (item) {
      return item.value === value;
    });
    selectedItem.onSelect();
  }

  function onPickerPresent() {
    if (blockActionsMenuPickerRef.current) {
      blockActionsMenuPickerRef.current.presentPicker();
    }
  }

  var disabledButtonIndices = options.map(function (option, index) {
    return option.disabled && index + 1;
  }).filter(Boolean);
  var accessibilityHint = _reactNative.Platform.OS === 'ios' ? (0, _i18n.__)('Double tap to open Action Sheet with available options') : (0, _i18n.__)('Double tap to open Bottom Sheet with available options');

  var getAnchor = function getAnchor() {
    return anchorNodeRef ? (0, _reactNative.findNodeHandle)(anchorNodeRef) : undefined;
  };

  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.ToolbarButton, {
    title: (0, _i18n.__)('Open Block Actions Menu'),
    onClick: onPickerPresent,
    icon: _icons.moreHorizontalMobile,
    extraProps: {
      hint: accessibilityHint
    }
  }), (0, _element.createElement)(_components.Picker, {
    ref: blockActionsMenuPickerRef,
    options: options,
    onChange: onPickerSelect,
    destructiveButtonIndex: options.length,
    disabledButtonIndices: disabledButtonIndices,
    hideCancelButton: _reactNative.Platform.OS !== 'ios',
    leftAlign: true,
    getAnchor: getAnchor // translators: %s: block title e.g: "Paragraph".
    ,
    title: (0, _i18n.sprintf)((0, _i18n.__)('%s block options'), blockTitle)
  }), (0, _element.createElement)(_blockTransformationsMenu.default, {
    anchorNodeRef: anchorNodeRef,
    blockTitle: blockTitle,
    pickerRef: blockTransformationMenuPickerRef,
    possibleTransformations: selectedBlockPossibleTransformations,
    selectedBlock: getBlocksByClientId(selectedBlockClientId),
    selectedBlockClientId: selectedBlockClientId
  }));
};

var _default = (0, _compose.compose)((0, _data.withSelect)(function (select, _ref2) {
  var clientIds = _ref2.clientIds;

  var _select = select(_store.store),
      getBlockIndex = _select.getBlockIndex,
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockOrder = _select.getBlockOrder,
      getBlockName = _select.getBlockName,
      getBlockTransformItems = _select.getBlockTransformItems,
      getBlock = _select.getBlock,
      getBlocksByClientId = _select.getBlocksByClientId,
      getSelectedBlockClientIds = _select.getSelectedBlockClientIds,
      canInsertBlockType = _select.canInsertBlockType;

  var normalizedClientIds = (0, _lodash.castArray)(clientIds);
  var block = getBlock(normalizedClientIds);
  var blockName = getBlockName(normalizedClientIds);
  var blockType = (0, _blocks.getBlockType)(blockName);
  var blockTitle = blockType === null || blockType === void 0 ? void 0 : blockType.title;
  var firstClientId = (0, _lodash.first)(normalizedClientIds);
  var rootClientId = getBlockRootClientId(firstClientId);
  var blockOrder = getBlockOrder(rootClientId);
  var firstIndex = getBlockIndex(firstClientId, rootClientId);
  var lastIndex = getBlockIndex((0, _lodash.last)(normalizedClientIds), rootClientId);
  var isDefaultBlock = blockName === (0, _blocks.getDefaultBlockName)();
  var isEmptyContent = (block === null || block === void 0 ? void 0 : block.attributes.content) === '';
  var isExactlyOneBlock = blockOrder.length === 1;
  var isEmptyDefaultBlock = isExactlyOneBlock && isDefaultBlock && isEmptyContent;
  var selectedBlockClientId = (0, _lodash.first)(getSelectedBlockClientIds());
  var selectedBlock = selectedBlockClientId ? (0, _lodash.first)(getBlocksByClientId(selectedBlockClientId)) : undefined;
  var selectedBlockPossibleTransformations = selectedBlock ? getBlockTransformItems([selectedBlock], rootClientId) : [];
  return {
    blockTitle: blockTitle,
    canInsertBlockType: canInsertBlockType,
    currentIndex: firstIndex,
    getBlocksByClientId: getBlocksByClientId,
    isEmptyDefaultBlock: isEmptyDefaultBlock,
    isFirst: firstIndex === 0,
    isLast: lastIndex === blockOrder.length - 1,
    rootClientId: rootClientId,
    selectedBlockClientId: selectedBlockClientId,
    selectedBlockPossibleTransformations: selectedBlockPossibleTransformations
  };
}), (0, _data.withDispatch)(function (dispatch, _ref3, _ref4) {
  var clientIds = _ref3.clientIds,
      rootClientId = _ref3.rootClientId,
      currentIndex = _ref3.currentIndex;
  var select = _ref4.select;

  var _dispatch = dispatch(_store.store),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp,
      duplicateBlocks = _dispatch.duplicateBlocks,
      removeBlocks = _dispatch.removeBlocks,
      insertBlock = _dispatch.insertBlock,
      replaceBlocks = _dispatch.replaceBlocks;

  var _dispatch2 = dispatch('core/edit-post'),
      _openGeneralSidebar = _dispatch2.openGeneralSidebar;

  var _select2 = select(_store.store),
      getBlockSelectionEnd = _select2.getBlockSelectionEnd,
      getBlock = _select2.getBlock;

  var _dispatch3 = dispatch(_notices.store),
      createSuccessNotice = _dispatch3.createSuccessNotice;

  return {
    createSuccessNotice: createSuccessNotice,
    duplicateBlock: function duplicateBlock() {
      return duplicateBlocks(clientIds);
    },
    onMoveDown: (0, _lodash.partial)(moveBlocksDown, clientIds, rootClientId),
    onMoveUp: (0, _lodash.partial)(moveBlocksUp, clientIds, rootClientId),
    openGeneralSidebar: function openGeneralSidebar() {
      return _openGeneralSidebar('edit-post/block');
    },
    pasteBlock: function pasteBlock(clipboardBlock) {
      var canReplaceBlock = (0, _blocks.isUnmodifiedDefaultBlock)(getBlock(getBlockSelectionEnd()));

      if (!canReplaceBlock) {
        var insertedBlock = (0, _blocks.createBlock)(clipboardBlock.name, clipboardBlock.attributes, clipboardBlock.innerBlocks);
        insertBlock(insertedBlock, currentIndex + 1, rootClientId);
      } else {
        replaceBlocks(clientIds, clipboardBlock);
      }
    },
    removeBlocks: removeBlocks
  };
}), _compose.withInstanceId)(BlockActionsMenu);

exports.default = _default;
