// @wordpress/block-library/build/table-of-contents/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _icon = _interopRequireDefault(require("./icon"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/table-of-contents",
  category: "layout",
  attributes: {
    onlyIncludeCurrentPage: {
      type: "boolean",
      "default": false
    }
  },
  usesContext: ["postId"],
  supports: {
    html: false
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Table of Contents'),
  description: (0, _i18n.__)('Summarize your post with a list of headings. Add HTML anchors to Heading blocks to link them here.'),
  icon: _icon.default,
  keywords: [(0, _i18n.__)('document outline'), (0, _i18n.__)('summary')],
  edit: _edit.default
};
exports.settings = settings;
