// @wordpress/block-library/build-module/query/edit/query-block-setup.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { cloneDeep } from 'lodash';
/**
 * WordPress dependencies
 */

import { useDispatch } from '@wordpress/data';
import { __, _x } from '@wordpress/i18n';
import { SelectControl, ToggleControl } from '@wordpress/components';
import { cloneBlock, createBlocksFromInnerBlocksTemplate } from '@wordpress/blocks';
import { store as blockEditorStore } from '@wordpress/block-editor';
/**
 * Internal dependencies
 */

import BlockSetup from './block-setup';
import { usePostTypes } from '../utils';

var QueryBlockSetup = function QueryBlockSetup(_ref) {
  var clientId = _ref.clientId,
      query = _ref.attributes.query,
      setAttributes = _ref.setAttributes,
      blockName = _ref.name;
  var postType = query.postType,
      inherit = query.inherit;

  var _useDispatch = useDispatch(blockEditorStore),
      replaceBlocks = _useDispatch.replaceBlocks,
      replaceInnerBlocks = _useDispatch.replaceInnerBlocks;

  var _usePostTypes = usePostTypes(),
      postTypesSelectOptions = _usePostTypes.postTypesSelectOptions;

  var updateQuery = function updateQuery(newQuery) {
    return setAttributes({
      query: _objectSpread(_objectSpread({}, query), newQuery)
    });
  };

  var onVariationSelect = function onVariationSelect(nextVariation) {
    if (nextVariation.attributes) {
      setAttributes(nextVariation.attributes);
    }

    if (nextVariation.innerBlocks) {
      replaceInnerBlocks(clientId, createBlocksFromInnerBlocksTemplate(nextVariation.innerBlocks), false);
    }
  };

  var onBlockPatternSelect = function onBlockPatternSelect(blocks) {
    var clonedBlocks = blocks.map(function (block) {
      var clone = cloneBlock(block);
      /**
       * TODO: this check will be revised with the ongoing work on block patterns.
       * For now we keep the value of posts per page (`query.perPage`) from Query patterns
       * so as to preview the pattern as intended, without possible big previews.
       * During insertion, we need to override the Query's attributes that can be set in
       * the Placeholder and we unset the `perPage` value to be set appropriately by Query block.
       */

      if (block.name === 'core/query') {
        /**
         * We need to `cloneDeep` the Query's attributes, as `cloneBlock` does a swallow
         * copy of the block.
         */
        var queryAttributes = cloneDeep(clone.attributes);
        Object.assign(queryAttributes.query, {
          inherit: query.inherit,
          postType: query.postType,
          perPage: null
        });
        return _objectSpread(_objectSpread({}, clone), {}, {
          attributes: queryAttributes
        });
      }

      return clone;
    });
    replaceBlocks(clientId, clonedBlocks);
  };

  var inheritToggleHelp = !!inherit ? _x('Inherit the global query depending on the URL.', 'Query block `inherit` option helping text') : _x('Customize the query arguments.', 'Query block `inherit` option helping text');
  return createElement(BlockSetup, {
    blockName: blockName,
    useLayoutSetup: true,
    onVariationSelect: onVariationSelect,
    onBlockPatternSelect: onBlockPatternSelect
  }, createElement("div", {
    className: "block-attributes-setup-container"
  }, createElement(ToggleControl, {
    label: __('Inherit query from URL'),
    checked: !!inherit,
    onChange: function onChange(value) {
      return updateQuery({
        inherit: !!value
      });
    },
    help: inheritToggleHelp
  }), !inherit && createElement(SelectControl, {
    options: postTypesSelectOptions,
    value: postType,
    label: __('Post Type'),
    onChange: function onChange(newValue) {
      return updateQuery({
        postType: newValue
      });
    }
  })));
};

export default QueryBlockSetup;
