// @wordpress/block-library/build/query/edit/block-setup/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _data = require("@wordpress/data");

var _blockEditor = require("@wordpress/block-editor");

var _blocks = require("@wordpress/blocks");

var _components = require("@wordpress/components");

var _layoutStep = _interopRequireDefault(require("./layout-step"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var BlockSetup = function BlockSetup(_ref) {
  var _blockType$icon;

  var blockName = _ref.blockName,
      useLayoutSetup = _ref.useLayoutSetup,
      _ref$onVariationSelec = _ref.onVariationSelect,
      onVariationSelect = _ref$onVariationSelec === void 0 ? function () {} : _ref$onVariationSelec,
      _ref$onBlockPatternSe = _ref.onBlockPatternSelect,
      onBlockPatternSelect = _ref$onBlockPatternSe === void 0 ? function () {} : _ref$onBlockPatternSe,
      children = _ref.children;

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select(_blocks.store),
        getBlockType = _select.getBlockType;

    return {
      blockType: getBlockType(blockName)
    };
  }, [blockName]),
      blockType = _useSelect.blockType;

  var blockProps = (0, _blockEditor.useBlockProps)();
  return (0, _element.createElement)("div", blockProps, (0, _element.createElement)(_components.Placeholder, {
    icon: blockType === null || blockType === void 0 ? void 0 : (_blockType$icon = blockType.icon) === null || _blockType$icon === void 0 ? void 0 : _blockType$icon.src,
    label: blockType === null || blockType === void 0 ? void 0 : blockType.title,
    isColumnLayout: true
  }, useLayoutSetup && (0, _element.createElement)(_layoutStep.default, {
    blockType: blockType,
    onVariationSelect: onVariationSelect,
    onBlockPatternSelect: onBlockPatternSelect
  }), children));
};

var _default = BlockSetup;
exports.default = _default;
