// @wordpress/block-library/build/query/edit/block-setup/layout-step.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _data = require("@wordpress/data");

var _blocks = require("@wordpress/blocks");

var _compose = require("@wordpress/compose");

var _blockEditor = require("@wordpress/block-editor");

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */
var LayoutSetupStep = function LayoutSetupStep(_ref) {
  var blockType = _ref.blockType,
      onVariationSelect = _ref.onVariationSelect,
      onBlockPatternSelect = _ref.onBlockPatternSelect;

  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      showBack = _useState2[0],
      setShowBack = _useState2[1];

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select(_blocks.store),
        getBlockVariations = _select.getBlockVariations,
        getDefaultBlockVariation = _select.getDefaultBlockVariation;

    var _select2 = select(_blockEditor.store),
        __experimentalGetPatternsByBlockTypes = _select2.__experimentalGetPatternsByBlockTypes;

    var name = blockType.name;

    var _patterns = __experimentalGetPatternsByBlockTypes(name);

    var _blockVariations = getBlockVariations(name, 'block');

    return {
      defaultVariation: getDefaultBlockVariation(name, 'block'),
      blockVariations: _blockVariations,
      patterns: _patterns,
      hasBlockVariations: !!(_blockVariations !== null && _blockVariations !== void 0 && _blockVariations.length)
    };
  }, [blockType]),
      defaultVariation = _useSelect.defaultVariation,
      blockVariations = _useSelect.blockVariations,
      patterns = _useSelect.patterns,
      hasBlockVariations = _useSelect.hasBlockVariations;

  var _useState3 = (0, _element.useState)(!(patterns !== null && patterns !== void 0 && patterns.length) && hasBlockVariations),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      showBlockVariations = _useState4[0],
      setShowBlockVariations = _useState4[1];

  var composite = (0, _components.__unstableUseCompositeState)(); // Show nothing if block variations and block pattterns do not exist.

  if (!hasBlockVariations && !(patterns !== null && patterns !== void 0 && patterns.length)) return null;
  var showPatternsList = !showBlockVariations && !!patterns.length;
  return (0, _element.createElement)(_element.Fragment, null, showBack && (0, _element.createElement)(_components.Button, {
    className: "block-setup-block-layout-back-button",
    icon: (0, _i18n.isRTL)() ? _icons.chevronRight : _icons.chevronLeft,
    isTertiary: true,
    onClick: function onClick() {
      setShowBack(false);
      setShowBlockVariations(false);
    }
  }, (0, _i18n.__)('Back')), (0, _element.createElement)(_components.__unstableComposite, (0, _extends2.default)({}, composite, {
    role: "listbox",
    className: "block-setup-block-layout-list__container",
    "aria-label": (0, _i18n.__)('Layout list')
  }), showBlockVariations && (0, _element.createElement)(_element.Fragment, null, blockVariations.map(function (variation) {
    return (0, _element.createElement)(BlockVariation, {
      key: variation.name,
      variation: variation,
      onSelect: function onSelect() {
        var nextVariation = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : defaultVariation;
        return onVariationSelect(nextVariation);
      },
      composite: composite
    });
  })), showPatternsList && (0, _element.createElement)(_element.Fragment, null, patterns.map(function (pattern) {
    return (0, _element.createElement)(BlockPattern, {
      key: pattern.name,
      pattern: pattern,
      onSelect: onBlockPatternSelect,
      composite: composite
    });
  })), !showBlockVariations && hasBlockVariations && (0, _element.createElement)(BlockVariation, {
    key: defaultVariation.name,
    variation: {
      title: (0, _i18n.__)('Start empty')
    },
    onSelect: function onSelect() {
      setShowBack(true);
      setShowBlockVariations(true);
    },
    composite: composite
  })));
};

function BlockPattern(_ref2) {
  var pattern = _ref2.pattern,
      onSelect = _ref2.onSelect,
      composite = _ref2.composite;
  var viewportWidth = pattern.viewportWidth,
      blocks = pattern.blocks;
  var descriptionId = (0, _compose.useInstanceId)(BlockPattern, 'block-setup-block-layout-list__item-description');
  return (0, _element.createElement)("div", {
    className: "block-setup-block-layout-list__list-item",
    "aria-label": pattern.title,
    "aria-describedby": pattern.description ? descriptionId : undefined
  }, (0, _element.createElement)(_components.__unstableCompositeItem, (0, _extends2.default)({
    role: "option",
    as: "div"
  }, composite, {
    className: "block-setup-block-layout-list__item",
    onClick: function onClick() {
      return onSelect(blocks);
    }
  }), (0, _element.createElement)(_blockEditor.BlockPreview, {
    blocks: blocks,
    viewportWidth: viewportWidth
  })), (0, _element.createElement)("div", {
    className: "block-setup-block-layout-list__item-title"
  }, pattern.title), !!pattern.description && (0, _element.createElement)(_components.VisuallyHidden, {
    id: descriptionId
  }, pattern.description));
}

function BlockVariation(_ref3) {
  var variation = _ref3.variation,
      onSelect = _ref3.onSelect,
      composite = _ref3.composite;
  var descriptionId = (0, _compose.useInstanceId)(BlockVariation, 'block-setup-block-layout-list__item-description');
  return (0, _element.createElement)("div", {
    className: "block-setup-block-layout-list__list-item is-block-variation",
    "aria-label": variation.title,
    "aria-describedby": variation.description ? descriptionId : undefined
  }, (0, _element.createElement)(_components.__unstableCompositeItem, (0, _extends2.default)({
    role: "option",
    as: "div"
  }, composite, {
    className: "block-setup-block-layout-list__item",
    onClick: function onClick() {
      return onSelect(variation);
    },
    label: variation.description || variation.title
  }), variation.icon && (0, _element.createElement)("div", {
    className: "block-setup-block-layout-list__item-variation-icon"
  }, (0, _element.createElement)(_components.Icon, {
    icon: variation.icon,
    size: 48
  }))), (0, _element.createElement)("div", {
    className: "block-setup-block-layout-list__item-title"
  }, variation.title), !!variation.description && (0, _element.createElement)(_components.VisuallyHidden, {
    id: descriptionId
  }, variation.description));
}

var _default = LayoutSetupStep;
exports.default = _default;
