// @wordpress/block-library/build-module/block/edit.native.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { ActivityIndicator, Platform, Text, TouchableWithoutFeedback, View } from 'react-native';
/**
 * WordPress dependencies
 */

import { useState } from '@wordpress/element';
import { useEntityBlockEditor, store as coreStore } from '@wordpress/core-data';
import { BottomSheet, Icon, Disabled } from '@wordpress/components';
import { useSelect } from '@wordpress/data';
import { __ } from '@wordpress/i18n';
import { BlockEditorProvider, BlockList, store as blockEditorStore } from '@wordpress/block-editor';
import { usePreferredColorSchemeStyle } from '@wordpress/compose';
import { help } from '@wordpress/icons';
import { store as reusableBlocksStore } from '@wordpress/reusable-blocks';
/**
 * Internal dependencies
 */

import styles from './editor.scss';
import EditTitle from './edit-title';
export default function ReusableBlockEdit(_ref) {
  var ref = _ref.attributes.ref,
      clientId = _ref.clientId,
      isSelected = _ref.isSelected;
  var recordArgs = ['postType', 'wp_block', ref];

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      showHelp = _useState2[0],
      setShowHelp = _useState2[1];

  var infoTextStyle = usePreferredColorSchemeStyle(styles.infoText, styles.infoTextDark);
  var infoTitleStyle = usePreferredColorSchemeStyle(styles.infoTitle, styles.infoTitleDark);
  var infoSheetIconStyle = usePreferredColorSchemeStyle(styles.infoSheetIcon, styles.infoSheetIconDark);
  var spinnerStyle = usePreferredColorSchemeStyle(styles.spinner, styles.spinnerDark);

  var _useSelect = useSelect(function (select) {
    var _select2, _select3;

    var _select = select(blockEditorStore),
        getSettings = _select.getSettings;

    return {
      reusableBlock: (_select2 = select(coreStore)).getEditedEntityRecord.apply(_select2, recordArgs),
      hasResolved: select(coreStore).hasFinishedResolution('getEditedEntityRecord', recordArgs),
      isSaving: (_select3 = select(coreStore)).isSavingEntityRecord.apply(_select3, recordArgs),
      canUserUpdate: select(coreStore).canUser('update', 'blocks', ref),
      isEditing: select(reusableBlocksStore).__experimentalIsEditingReusableBlock(clientId),
      settings: getSettings()
    };
  }, [ref, clientId]),
      reusableBlock = _useSelect.reusableBlock,
      hasResolved = _useSelect.hasResolved,
      isEditing = _useSelect.isEditing,
      settings = _useSelect.settings;

  var _useEntityBlockEditor = useEntityBlockEditor('postType', 'wp_block', {
    id: ref
  }),
      _useEntityBlockEditor2 = _slicedToArray(_useEntityBlockEditor, 3),
      blocks = _useEntityBlockEditor2[0],
      onInput = _useEntityBlockEditor2[1],
      onChange = _useEntityBlockEditor2[2];

  function openSheet() {
    setShowHelp(true);
  }

  function closeSheet() {
    setShowHelp(false);
  }

  function renderSheet() {
    var infoTitle = Platform.OS === 'android' ? __("Reusable blocks aren't editable on WordPress for Android") : __("Reusable blocks aren't editable on WordPress for iOS");
    return createElement(BottomSheet, {
      isVisible: showHelp,
      hideHeader: true,
      onClose: closeSheet
    }, createElement(View, {
      style: styles.infoContainer
    }, createElement(Icon, {
      icon: help,
      color: infoSheetIconStyle.color,
      size: styles.infoSheetIcon.size
    }), createElement(Text, {
      style: [infoTextStyle, infoTitleStyle]
    }, infoTitle)));
  }

  if (!hasResolved) {
    return createElement(View, {
      style: spinnerStyle
    }, createElement(ActivityIndicator, {
      animating: true
    }));
  }

  if (!reusableBlock) {
    return createElement(Text, null, __('Block has been deleted or is unavailable.'));
  }

  var title = reusableBlock.title;
  var element = createElement(BlockEditorProvider, {
    settings: settings,
    value: blocks,
    onChange: onChange,
    onInput: onInput
  }, createElement(BlockList, {
    withFooter: false,
    marginHorizontal: 0
  }));

  if (!isEditing) {
    element = createElement(Disabled, null, element);
  }

  return createElement(TouchableWithoutFeedback, {
    disabled: !isSelected,
    accessibilityLabel: __('Help button'),
    accessibilityRole: 'button',
    accessibilityHint: __('Tap here to show help'),
    onPress: openSheet
  }, createElement(View, null, isSelected && createElement(EditTitle, {
    title: title
  }), element, renderSheet()));
}
