// @wordpress/block-library/build-module/template-part/edit/placeholder/index.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import { createElement, Fragment } from "@wordpress/element";
import _regeneratorRuntime from "@babel/runtime/regenerator";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { useCallback } from '@wordpress/element';
import { useDispatch } from '@wordpress/data';
import { Placeholder, Dropdown, Button } from '@wordpress/components';
import { blockDefault } from '@wordpress/icons';
import { serialize } from '@wordpress/blocks';
import { store as coreStore } from '@wordpress/core-data';
/**
 * Internal dependencies
 */

import TemplatePartSelection from '../selection';
export default function TemplatePartPlaceholder(_ref) {
  var area = _ref.area,
      setAttributes = _ref.setAttributes,
      innerBlocks = _ref.innerBlocks;

  var _useDispatch = useDispatch(coreStore),
      saveEntityRecord = _useDispatch.saveEntityRecord;

  var onCreate = useCallback( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
    var title, record, templatePart;
    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            title = __('Untitled Template Part'); // If we have `area` set from block attributes, means an exposed
            // block variation was inserted. So add this prop to the template
            // part entity on creation. Afterwards remove `area` value from
            // block attributes.

            record = {
              title: title,
              slug: 'template-part',
              content: serialize(innerBlocks),
              // `area` is filterable on the server and defaults to `UNCATEGORIZED`
              // if provided value is not allowed.
              area: area
            };
            _context.next = 4;
            return saveEntityRecord('postType', 'wp_template_part', record);

          case 4:
            templatePart = _context.sent;
            setAttributes({
              slug: templatePart.slug,
              theme: templatePart.theme,
              area: undefined
            });

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  })), [setAttributes, area]);
  return createElement(Placeholder, {
    icon: blockDefault,
    label: __('Template Part'),
    instructions: __('Create a new template part or pick an existing one from the list.')
  }, createElement(Dropdown, {
    contentClassName: "wp-block-template-part__placeholder-preview-dropdown-content",
    position: "bottom right left",
    renderToggle: function renderToggle(_ref3) {
      var isOpen = _ref3.isOpen,
          onToggle = _ref3.onToggle;
      return createElement(Fragment, null, createElement(Button, {
        isPrimary: true,
        onClick: onToggle,
        "aria-expanded": isOpen
      }, __('Choose existing')), createElement(Button, {
        isTertiary: true,
        onClick: onCreate
      }, __('New template part')));
    },
    renderContent: function renderContent(_ref4) {
      var onClose = _ref4.onClose;
      return createElement(TemplatePartSelection, {
        setAttributes: setAttributes,
        onClose: onClose
      });
    }
  }));
}
