// @wordpress/block-library/build-module/social-links/deprecated.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import classNames from 'classnames';
/**
 * WordPress dependencies
 */

import { InnerBlocks, useBlockProps } from '@wordpress/block-editor'; // Social Links block deprecations.

var deprecated = [// V1. Remove CSS variable use for colors.
{
  attributes: {
    iconColor: {
      type: 'string'
    },
    customIconColor: {
      type: 'string'
    },
    iconColorValue: {
      type: 'string'
    },
    iconBackgroundColor: {
      type: 'string'
    },
    customIconBackgroundColor: {
      type: 'string'
    },
    iconBackgroundColorValue: {
      type: 'string'
    },
    openInNewTab: {
      type: 'boolean',
      default: false
    },
    size: {
      type: 'string'
    }
  },
  providesContext: {
    openInNewTab: 'openInNewTab'
  },
  supports: {
    align: ['left', 'center', 'right'],
    anchor: true
  },
  save: function save(props) {
    var _props$attributes = props.attributes,
        iconBackgroundColorValue = _props$attributes.iconBackgroundColorValue,
        iconColorValue = _props$attributes.iconColorValue,
        itemsJustification = _props$attributes.itemsJustification,
        size = _props$attributes.size;
    var className = classNames(size, _defineProperty({
      'has-icon-color': iconColorValue,
      'has-icon-background-color': iconBackgroundColorValue
    }, "items-justified-".concat(itemsJustification), itemsJustification));
    var style = {
      '--wp--social-links--icon-color': iconColorValue,
      '--wp--social-links--icon-background-color': iconBackgroundColorValue
    };
    return createElement("ul", useBlockProps.save({
      className: className,
      style: style
    }), createElement(InnerBlocks.Content, null));
  }
}];
export default deprecated;
