// @wordpress/block-library/build/query-title/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _variations = _interopRequireDefault(require("./variations"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/query-title",
  category: "design",
  attributes: {
    type: {
      type: "string"
    },
    textAlign: {
      type: "string"
    },
    level: {
      type: "number",
      "default": 1
    }
  },
  supports: {
    align: ["wide", "full"],
    html: false,
    color: {
      gradients: true
    },
    fontSize: true,
    lineHeight: true,
    __experimentalFontFamily: true,
    __experimentalSelector: {
      "core/query-title/h1": {
        title: "h1",
        selector: "h1.wp-block-query-title",
        attributes: {
          level: 1
        }
      },
      "core/query-title/h2": {
        title: "h2",
        selector: "h2.wp-block-query-title",
        attributes: {
          level: 2
        }
      },
      "core/query-title/h3": {
        title: "h3",
        selector: "h3.wp-block-query-title",
        attributes: {
          level: 3
        }
      },
      "core/query-title/h4": {
        title: "h4",
        selector: "h4.wp-block-query-title",
        attributes: {
          level: 4
        }
      },
      "core/query-title/h5": {
        title: "h5",
        selector: "h5.wp-block-query-title",
        attributes: {
          level: 5
        }
      },
      "core/query-title/h6": {
        title: "h6",
        selector: "h6.wp-block-query-title",
        attributes: {
          level: 6
        }
      }
    }
  },
  editorStyle: "wp-block-query-title-editor"
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n.__)('Query Title'),
  description: (0, _i18n.__)('Display the query title.'),
  edit: _edit.default,
  variations: _variations.default
};
exports.settings = settings;
