// @wordpress/block-library/build/query-pagination-previous/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = QueryPaginationPreviousEdit;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

/**
 * WordPress dependencies
 */
function QueryPaginationPreviousEdit(_ref) {
  var label = _ref.attributes.label,
      setAttributes = _ref.setAttributes;
  return (0, _element.createElement)(_blockEditor.PlainText, (0, _extends2.default)({
    __experimentalVersion: 2,
    tagName: "a",
    style: {
      display: 'inline-block'
    },
    "aria-label": (0, _i18n.__)('Previous page link'),
    placeholder: (0, _i18n.__)('Previous Page'),
    value: label,
    onChange: function onChange(newLabel) {
      return setAttributes({
        label: newLabel
      });
    }
  }, (0, _blockEditor.useBlockProps)()));
}
