// @wordpress/components/build/date-time/date.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _moment = _interopRequireDefault(require("moment"));

var _classnames = _interopRequireDefault(require("classnames"));

var _DayPickerSingleDateController = _interopRequireDefault(require("react-dates/lib/components/DayPickerSingleDateController"));

var _i18n = require("@wordpress/i18n");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * Module Constants
 */
var TIMEZONELESS_FORMAT = 'YYYY-MM-DDTHH:mm:ss';
var ARIAL_LABEL_TIME_FORMAT = 'dddd, LL';

function DatePickerDay(_ref) {
  var day = _ref.day,
      _ref$events = _ref.events,
      events = _ref$events === void 0 ? [] : _ref$events;
  var ref = (0, _element.useRef)();
  /*
   * a11y hack to make the `There is/are n events` string
   * available speaking for readers,
   * re-defining the aria-label attribute.
   * This attribute is handled by the react-dates component.
   */

  (0, _element.useEffect)(function () {
    var _ref$current;

    // Bail when no parent node.
    if (!(ref !== null && ref !== void 0 && (_ref$current = ref.current) !== null && _ref$current !== void 0 && _ref$current.parentNode)) {
      return;
    }

    var parentNode = ref.current.parentNode;
    var dayAriaLabel = (0, _moment.default)(day).format(ARIAL_LABEL_TIME_FORMAT);

    if (!events.length) {
      // Set aria-label without event description.
      parentNode.setAttribute('aria-label', dayAriaLabel);
      return;
    }

    var dayWithEventsDescription = (0, _i18n.sprintf)( // translators: 1: Calendar day format, 2: Calendar event number.
    (0, _i18n._n)('%1$s. There is %2$d event.', '%1$s. There are %2$d events.', events.length), dayAriaLabel, events.length);
    parentNode.setAttribute('aria-label', dayWithEventsDescription);
  }, [events.length]);
  return (0, _element.createElement)("div", {
    ref: ref,
    className: (0, _classnames.default)('components-datetime__date__day', {
      'has-events': events === null || events === void 0 ? void 0 : events.length
    })
  }, day.format('D'));
}

var DatePicker = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(DatePicker, _Component);

  var _super = _createSuper(DatePicker);

  function DatePicker() {
    var _this;

    (0, _classCallCheck2.default)(this, DatePicker);
    _this = _super.apply(this, arguments);
    _this.onChangeMoment = _this.onChangeMoment.bind((0, _assertThisInitialized2.default)(_this));
    _this.nodeRef = (0, _element.createRef)();
    _this.onMonthPreviewedHandler = _this.onMonthPreviewedHandler.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(DatePicker, [{
    key: "onMonthPreviewedHandler",
    value: function onMonthPreviewedHandler(newMonthDate) {
      var _this$props;

      (_this$props = this.props) === null || _this$props === void 0 ? void 0 : _this$props.onMonthPreviewed(newMonthDate.toISOString());
      this.keepFocusInside();
    }
    /*
     * Todo: We should remove this function ASAP.
     * It is kept because focus is lost when we click on the previous and next month buttons.
     * This focus loss closes the date picker popover.
     * Ideally we should add an upstream commit on react-dates to fix this issue.
     */

  }, {
    key: "keepFocusInside",
    value: function keepFocusInside() {
      if (!this.nodeRef.current) {
        return;
      }

      var ownerDocument = this.nodeRef.current.ownerDocument;
      var activeElement = ownerDocument.activeElement; // If focus was lost.

      if (!activeElement || !this.nodeRef.current.contains(ownerDocument.activeElement)) {
        // Retrieve the focus region div.
        var focusRegion = this.nodeRef.current.querySelector('.DayPicker_focusRegion');

        if (!focusRegion) {
          return;
        } // Keep the focus on focus region.


        focusRegion.focus();
      }
    }
  }, {
    key: "onChangeMoment",
    value: function onChangeMoment(newDate) {
      var _this$props2 = this.props,
          currentDate = _this$props2.currentDate,
          onChange = _this$props2.onChange; // If currentDate is null, use now as momentTime to designate hours, minutes, seconds.

      var momentDate = currentDate ? (0, _moment.default)(currentDate) : (0, _moment.default)();
      var momentTime = {
        hours: momentDate.hours(),
        minutes: momentDate.minutes(),
        seconds: 0
      };
      onChange(newDate.set(momentTime).format(TIMEZONELESS_FORMAT)); // Keep focus on the date picker.

      this.keepFocusInside();
    }
    /**
     * Create a Moment object from a date string. With no currentDate supplied, default to a Moment
     * object representing now. If a null value is passed, return a null value.
     *
     * @param {?string} currentDate Date representing the currently selected date or null to signify no selection.
     * @return {?moment.Moment} Moment object for selected date or null.
     */

  }, {
    key: "getMomentDate",
    value: function getMomentDate(currentDate) {
      if (null === currentDate) {
        return null;
      }

      return currentDate ? (0, _moment.default)(currentDate) : (0, _moment.default)();
    }
  }, {
    key: "getEventsPerDay",
    value: function getEventsPerDay(day) {
      var _this$props$events;

      if (!((_this$props$events = this.props.events) !== null && _this$props$events !== void 0 && _this$props$events.length)) {
        return [];
      }

      return this.props.events.filter(function (eventDay) {
        return day.isSame(eventDay.date, 'day');
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props3 = this.props,
          currentDate = _this$props3.currentDate,
          isInvalidDate = _this$props3.isInvalidDate;
      var momentDate = this.getMomentDate(currentDate);
      return (0, _element.createElement)("div", {
        className: "components-datetime__date",
        ref: this.nodeRef
      }, (0, _element.createElement)(_DayPickerSingleDateController.default, {
        date: momentDate,
        daySize: 30,
        focused: true,
        hideKeyboardShortcutsPanel: true // This is a hack to force the calendar to update on month or year change
        // https://github.com/airbnb/react-dates/issues/240#issuecomment-361776665
        ,
        key: "datepicker-controller-".concat(momentDate ? momentDate.format('MM-YYYY') : 'null'),
        noBorder: true,
        numberOfMonths: 1,
        onDateChange: this.onChangeMoment,
        transitionDuration: 0,
        weekDayFormat: "ddd",
        dayAriaLabelFormat: ARIAL_LABEL_TIME_FORMAT,
        isRTL: (0, _i18n.isRTL)(),
        isOutsideRange: function isOutsideRange(date) {
          return isInvalidDate && isInvalidDate(date.toDate());
        },
        onPrevMonthClick: this.onMonthPreviewedHandler,
        onNextMonthClick: this.onMonthPreviewedHandler,
        renderDayContents: function renderDayContents(day) {
          return (0, _element.createElement)(DatePickerDay, {
            day: day,
            events: _this2.getEventsPerDay(day)
          });
        }
      }));
    }
  }]);
  return DatePicker;
}(_element.Component);

var _default = DatePicker;
exports.default = _default;
