// @wordpress/block-library/build-module/buttons/edit.native.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import { debounce } from 'lodash';
import { View } from 'react-native';
/**
 * WordPress dependencies
 */

import { BlockControls, InnerBlocks, JustifyContentControl, store as blockEditorStore } from '@wordpress/block-editor';
import { createBlock } from '@wordpress/blocks';
import { useResizeObserver } from '@wordpress/compose';
import { useDispatch, useSelect } from '@wordpress/data';
import { useState, useEffect, useRef, useCallback } from '@wordpress/element';
import { alignmentHelpers } from '@wordpress/components';
/**
 * Internal dependencies
 */

import { name as buttonBlockName } from '../button/';
import styles from './editor.scss';
var ALLOWED_BLOCKS = [buttonBlockName];
var BUTTONS_TEMPLATE = [['core/button']];
var layoutProp = {
  type: 'default',
  alignments: []
};
export default function ButtonsEdit(_ref) {
  var _ref$attributes = _ref.attributes,
      contentJustification = _ref$attributes.contentJustification,
      align = _ref$attributes.align,
      clientId = _ref.clientId,
      isSelected = _ref.isSelected,
      setAttributes = _ref.setAttributes,
      blockWidth = _ref.blockWidth;

  var _useResizeObserver = useResizeObserver(),
      _useResizeObserver2 = _slicedToArray(_useResizeObserver, 2),
      resizeObserver = _useResizeObserver2[0],
      sizes = _useResizeObserver2[1];

  var _useState = useState(0),
      _useState2 = _slicedToArray(_useState, 2),
      maxWidth = _useState2[0],
      setMaxWidth = _useState2[1];

  var spacing = styles.spacing.marginLeft;

  var _useSelect = useSelect(function (select) {
    var _select = select(blockEditorStore),
        getBlockCount = _select.getBlockCount,
        getBlockParents = _select.getBlockParents,
        getSelectedBlockClientId = _select.getSelectedBlockClientId;

    var selectedBlockClientId = getSelectedBlockClientId();
    var selectedBlockParents = getBlockParents(selectedBlockClientId, true);
    return {
      isInnerButtonSelected: selectedBlockParents[0] === clientId,
      // The purpose of `shouldDelete` check is giving the ability to
      // pass to mobile toolbar function called `onDelete` which removes
      // the whole `Buttons` container along with the last inner button
      // when there is exactly one button.
      shouldDelete: getBlockCount(clientId) === 1
    };
  }, [clientId]),
      isInnerButtonSelected = _useSelect.isInnerButtonSelected,
      shouldDelete = _useSelect.shouldDelete;

  var _useSelect2 = useSelect(blockEditorStore),
      getBlockOrder = _useSelect2.getBlockOrder;

  var _useDispatch = useDispatch(blockEditorStore),
      insertBlock = _useDispatch.insertBlock,
      removeBlock = _useDispatch.removeBlock,
      selectBlock = _useDispatch.selectBlock;

  useEffect(function () {
    var _ref2 = sizes || {},
        width = _ref2.width;

    var isFullWidth = alignmentHelpers.isFullWidth;

    if (width) {
      var isFullWidthBlock = isFullWidth(align);
      setMaxWidth(isFullWidthBlock ? blockWidth : width);
    }
  }, [sizes, align]);
  var onAddNextButton = useCallback(debounce(function (selectedId) {
    var order = getBlockOrder(clientId);
    var selectedButtonIndex = order.findIndex(function (i) {
      return i === selectedId;
    });
    var index = selectedButtonIndex === -1 ? order.length + 1 : selectedButtonIndex;
    var insertedBlock = createBlock('core/button');
    insertBlock(insertedBlock, index, clientId);
    selectBlock(insertedBlock.clientId);
  }, 200), []);
  var renderFooterAppender = useRef(function () {
    return createElement(View, {
      style: styles.appenderContainer
    }, createElement(InnerBlocks.ButtonBlockAppender, {
      isFloating: true,
      onAddBlock: onAddNextButton
    }));
  });
  var justifyControls = ['left', 'center', 'right'];
  var remove = useCallback(function () {
    return removeBlock(clientId);
  }, [clientId]);
  var shouldRenderFooterAppender = isSelected || isInnerButtonSelected;
  return createElement(Fragment, null, isSelected && createElement(BlockControls, {
    group: "block"
  }, createElement(JustifyContentControl, {
    allowedControls: justifyControls,
    value: contentJustification,
    onChange: function onChange(value) {
      return setAttributes({
        contentJustification: value
      });
    },
    popoverProps: {
      position: 'bottom right',
      isAlternate: true
    }
  })), resizeObserver, createElement(InnerBlocks, {
    allowedBlocks: ALLOWED_BLOCKS,
    template: BUTTONS_TEMPLATE,
    renderFooterAppender: shouldRenderFooterAppender && renderFooterAppender.current,
    orientation: "horizontal",
    horizontalAlignment: contentJustification,
    onDeleteBlock: shouldDelete ? remove : undefined,
    onAddBlock: onAddNextButton,
    parentWidth: maxWidth // This value controls the width of that the buttons are able to expand to.
    ,
    marginHorizontal: spacing,
    marginVertical: spacing,
    __experimentalLayout: layoutProp,
    templateInsertUpdatesSelection: true,
    blockWidth: blockWidth
  }));
}
