// @wordpress/block-library/build-module/buttons/edit.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { BlockControls, useBlockProps, __experimentalUseInnerBlocksProps as useInnerBlocksProps, JustifyContentControl } from '@wordpress/block-editor';
/**
 * Internal dependencies
 */

import { name as buttonBlockName } from '../button';
var ALLOWED_BLOCKS = [buttonBlockName];
var BUTTONS_TEMPLATE = [['core/button']];

function ButtonsEdit(_ref) {
  var _classnames;

  var _ref$attributes = _ref.attributes,
      contentJustification = _ref$attributes.contentJustification,
      orientation = _ref$attributes.orientation,
      setAttributes = _ref.setAttributes;
  var blockProps = useBlockProps({
    className: classnames((_classnames = {}, _defineProperty(_classnames, "is-content-justification-".concat(contentJustification), contentJustification), _defineProperty(_classnames, 'is-vertical', orientation === 'vertical'), _classnames))
  });
  var innerBlocksProps = useInnerBlocksProps(blockProps, {
    allowedBlocks: ALLOWED_BLOCKS,
    template: BUTTONS_TEMPLATE,
    orientation: orientation,
    __experimentalLayout: {
      type: 'default',
      alignments: []
    },
    templateInsertUpdatesSelection: true
  });
  var justifyControls = orientation === 'vertical' ? ['left', 'center', 'right'] : ['left', 'center', 'right', 'space-between'];
  return createElement(Fragment, null, createElement(BlockControls, {
    group: "block"
  }, createElement(JustifyContentControl, {
    allowedControls: justifyControls,
    value: contentJustification,
    onChange: function onChange(value) {
      return setAttributes({
        contentJustification: value
      });
    },
    popoverProps: {
      position: 'bottom right',
      isAlternate: true
    }
  })), createElement("div", innerBlocksProps));
}

export default ButtonsEdit;
