// @wordpress/block-library/build/term-description/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _edit = _interopRequireDefault(require("./edit"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/term-description",
  category: "design",
  attributes: {
    textAlign: {
      type: "string"
    }
  },
  supports: {
    align: ["wide", "full"],
    html: false,
    fontSize: true,
    lineHeight: true,
    color: {
      link: true
    }
  },
  editorStyle: "wp-block-term-description-editor"
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Term Description', 'block title'),
  description: (0, _i18n.__)('Display the description of categories, tags and custom taxonomies when viewing an archive.'),
  icon: _icons.termDescription,
  edit: _edit.default
};
exports.settings = settings;
