// @wordpress/editor/build-module/components/provider/use-block-editor-settings.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { pick, defaultTo } from 'lodash';
/**
 * WordPress dependencies
 */

import { Platform, useMemo } from '@wordpress/element';
import { useDispatch, useSelect } from '@wordpress/data';
import { store as coreStore, __experimentalFetchLinkSuggestions as fetchLinkSuggestions } from '@wordpress/core-data';
/**
 * Internal dependencies
 */

import { mediaUpload } from '../../utils';
import { store as editorStore } from '../../store';
/**
 * React hook used to compute the block editor settings to use for the post editor.
 *
 * @param {Object}  settings    EditorProvider settings prop.
 * @param {boolean} hasTemplate Whether template mode is enabled.
 *
 * @return {Object} Block Editor Settings.
 */

function useBlockEditorSettings(settings, hasTemplate) {
  var _useSelect = useSelect(function (select) {
    var _select = select(editorStore),
        canUserUseUnfilteredHTML = _select.canUserUseUnfilteredHTML,
        isPostTitleSelected = _select.isPostTitleSelected;

    var _select2 = select(coreStore),
        canUser = _select2.canUser;

    return {
      canUseUnfilteredHTML: canUserUseUnfilteredHTML(),
      reusableBlocks: select(coreStore).getEntityRecords('postType', 'wp_block',
      /**
       * Unbounded queries are not supported on native so as a workaround, we set per_page with the maximum value that native version can handle.
       * Related issue: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2661
       */
      {
        per_page: Platform.select({
          web: -1,
          native: 10
        })
      }),
      hasUploadPermissions: defaultTo(canUser('create', 'media'), true),
      // This selector is only defined on mobile.
      isTitleSelected: isPostTitleSelected && isPostTitleSelected()
    };
  }, []),
      reusableBlocks = _useSelect.reusableBlocks,
      hasUploadPermissions = _useSelect.hasUploadPermissions,
      canUseUnfilteredHTML = _useSelect.canUseUnfilteredHTML,
      isTitleSelected = _useSelect.isTitleSelected;

  var _useDispatch = useDispatch(editorStore),
      undo = _useDispatch.undo;

  return useMemo(function () {
    return _objectSpread(_objectSpread({}, pick(settings, ['__experimentalBlockDirectory', '__experimentalBlockPatternCategories', '__experimentalBlockPatterns', '__experimentalFeatures', '__experimentalGlobalStylesBaseStyles', '__experimentalGlobalStylesUserEntityId', '__experimentalPreferredStyleVariations', '__experimentalSetIsInserterOpened', 'alignWide', 'allowedBlockTypes', 'bodyPlaceholder', 'codeEditingEnabled', 'colors', 'disableCustomColors', 'disableCustomFontSizes', 'disableCustomGradients', 'enableCustomLineHeight', 'enableCustomSpacing', 'enableCustomUnits', 'focusMode', 'fontSizes', 'gradients', 'hasFixedToolbar', 'hasReducedUI', 'imageDefaultSize', 'imageDimensions', 'imageEditing', 'imageSizes', 'isRTL', 'keepCaretInsideBlock', 'maxWidth', 'onUpdateDefaultBlockStyles', 'styles', 'template', 'templateLock', 'titlePlaceholder', 'supportsLayout', 'widgetTypesToHideFromLegacyWidgetBlock'])), {}, {
      mediaUpload: hasUploadPermissions ? mediaUpload : undefined,
      __experimentalReusableBlocks: reusableBlocks,
      __experimentalFetchLinkSuggestions: function __experimentalFetchLinkSuggestions(search, searchOptions) {
        return fetchLinkSuggestions(search, searchOptions, settings);
      },
      __experimentalCanUserUseUnfilteredHTML: canUseUnfilteredHTML,
      __experimentalUndo: undo,
      __experimentalShouldInsertAtTheTop: isTitleSelected,
      outlineMode: hasTemplate
    });
  }, [settings, hasUploadPermissions, reusableBlocks, canUseUnfilteredHTML, undo, isTitleSelected, hasTemplate]);
}

export default useBlockEditorSettings;
