// @wordpress/block-library/build/post-comments-link/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _icons = require("@wordpress/icons");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/post-comments-link",
  category: "design",
  usesContext: ["postType", "postId"],
  attributes: {
    textAlign: {
      type: "string"
    }
  },
  supports: {
    html: false,
    fontSize: true,
    color: {
      link: true,
      text: false
    },
    lineHeight: true
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Post Comments Link', 'block title'),
  description: (0, _i18n.__)('Displays the link to the current post comments.'),
  edit: _edit.default,
  icon: _icons.postCommentsCount
};
exports.settings = settings;
